---
title: "RNA-seq from 4 cell populations"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---


<!---

# cd /hpc/users/hoffmg01/work/cmc_analysis/src
cd /hpc/users/hoffmg01/work/CMC/covarr-de/deconvolution/fans4
# rm -rf molec_profiling_cache/
ml git pandoc
git pull
export OMP_NUM_THREADS=2
R

system("git pull")
rmarkdown::render("molec_profiling.Rmd", output_dir='./', intermediates_dir='./')


# TODO 
replace with FilterExpr()
add single cell
Add CMC bulk RNA-seq





# http
# https://hoffmg01.u.hpc.mssm.edu/molec_profiling.html

cd /Users/gabrielhoffman/workspace/repos/CMC/covarr-de/deconvolution/fans4
R


rmarkdown::render("molec_profiling.Rmd", output_dir='./', intermediates_dir='./')


--->


```{r knit2synapse, eval=FALSE}
library(synapser)
library(knit2synapse)
synLogin()
knit2synapse::createAndKnitToFolderEntity(file = "deconvolution/fans4/molec_profiling.Rmd",
                                          parentId ="syn22231629",
                                          folderName = 'FANS4 processing and testing',
                                          overwrite=TRUE, 
                                          knitmd=TRUE)
```


```{r load.packages, echo=FALSE, message=FALSE, results='hide', cache=FALSE}
suppressPackageStartupMessages({
library(variancePartition)
library(limma)
library(edgeR)
library(mvIC)
library(reshape2)
library(qvalue)
library(data.table)
library(BiocParallel)
library(knitr)
library(org.Hs.eg.db)
library(kableExtra)
library(ggplot2)
library(cowplot)
library(scran)
library(SummarizedExperiment)
library(gridExtra)
library(openxlsx)
library(synapser)
library(metafor)
library(qqman)
library(zenith)
library(UpSetR)
library(nnet)
library(PRROC)
library(githubr)
synLogin()
library(EnsDb.Hsapiens.v86)
detach("package:EnsDb.Hsapiens.v86", unload=TRUE)
library(EnsDb.Hsapiens.v86)
})

source("../../common_functions/common_functions.R")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  cache.lazy = FALSE,  dev = c("png", "pdf"), 
  fig.width=7, fig.height=7)

options(markdown.HTML.stylesheet = 'css/custom.css')
```

```{r synapse.parameters, include=FALSE, cache=TRUE, eval=TRUE}
thisFileName <- 'molec_profiling.Rmd'
# thisFile = thisFileName
# Github link
thisRepo <- getRepo(repository = "CommonMindConsortium/covarr-de", ref="branch", refName='master')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('deconvolution/fans4/',thisFileName))
```

```{r}
CODE <- Folder(name = "results", parentId = "syn26433312")
CODE <- synStore(CODE)

# Store data 
storeData = function( obj, file, description){
  saveRDS( obj, file=file)

  nEXP_OBJ = File(file, 
                  name = description, 
                  parentId = CODE$properties$id)
  synStore(nEXP_OBJ, used = ALL_USED_IDs, executed = thisFile)
}
```

### Reading RNA-seq data
```{r read.data, message=TRUE}
Ids = c('syn22321012', 'syn22321013', 'syn22321014', 'syn22321016')

# Read in RNA-seq featureCounts
###############################

# Phase 1
file = synGet('syn22321012')$path
df_featureCounts_gene = fread( file )

# Phase 2+3
# syn26000845 is really phase2 again
ids = c('syn26000845','syn25976117', 'syn25976127')[-1]
df_read = lapply(ids, function(id){
  df = fread(synGet(id)$path)
  colnames(df) = gsub("^Sample_", '', colnames(df))
  colnames(df) = gsub("-", '.', colnames(df))
  df
  })

# save batch info
df_batch = data.frame(Name = colnames(df_featureCounts_gene)[-c(1:6)], Batch=1)
df_batch = rbind(df_batch, data.frame(Name = colnames(df_read[[1]][,-c(1:6)]), Batch=2))
df_batch = rbind(df_batch, data.frame(Name = colnames(df_read[[2]][,-c(1:6)]), Batch=3))
df_batch$Batch = as.character(df_batch$Batch)

# sapply(df_read, ncol)
df_featureCounts_gene = cbind(df_featureCounts_gene, 
                          df_read[[1]][,-c(1:6)], 
                          df_read[[2]][,-c(1:6)])

geneInfo = df_featureCounts_gene[,1:6]
geneInfo$EnsId = trim_ensembl_ids( geneInfo$Geneid )
geneInfo = geneInfo %>% getGeneSymbol( 'Geneid' )

geneCounts = df_featureCounts_gene[,7:ncol(df_featureCounts_gene)]
geneCounts = as.matrix(geneCounts)
rownames(geneCounts) = geneInfo$Geneid 

# identify samples with improper names
idx = grep("Undetermined", colnames(geneCounts))
message("Dropping samples with improper names:\n",  paste(colnames(geneCounts)[idx], collapse="\n"))
geneCounts = geneCounts[,-idx]

idx = which(colSums(geneCounts) < 1e5)
message("Dropping samples with zero read counts:\n",  paste(colnames(geneCounts)[idx], collapse="\n"))
geneCounts = geneCounts[,-idx]

# remove genes with duplicated EnsId
exclude = duplicated(geneInfo$EnsId)
geneInfo = geneInfo[!exclude,]
geneCounts = geneCounts[!exclude,]
rownames(geneCounts) = trim_ensembl_ids( rownames(geneCounts) )

# convert to HUGO names
# df_hugo = select(org.Hs.eg.db, geneInfo$EnsId, "SYMBOL", "ENSEMBL")
# geneInfo = merge(geneInfo, df_hugo, by.x='EnsId', by.y="ENSEMBL")


```


```{r metadata}
# file = "/sc/arion/projects/CommonMind/roussp01a/MOLECULAR_PROFILING/meta-files/MolecularProfiling_MSSM_120samples_05.24.17-4.xlsx"
file = synGet('syn22321013')$path
# focused on this experiment
metadata1 = read.xlsx( file )

# from Synapse
# file = "/sc/arion/projects/CommonMind/roussp01a/MOLECULAR_PROFILING/meta-files/CMC_METADATA.csv"
file = synGet('syn22321014')$path
metadataCMC = fread( file )
```

### RNA-seq QC metrics
```{r QC, fig.width=18, fig.height=74}
# Phase 1 
# file = '/sc/arion/projects/CommonMind/roussp01a/MOLECULAR_PROFILING/rnaseq/step1-merged/QCmetrics/qc_metrics.RDS'
file = synGet('syn22321016')$path
qc_metrics.in = t(readRDS( file ))
colnames(qc_metrics.in) = qc_metrics.in[1,]
qc_metrics.in = as.data.frame(qc_metrics.in[-c(1:2),], stringsAsFactors=FALSE)
qc_metrics = apply(qc_metrics.in, 2, as.numeric)
rownames(qc_metrics) = gsub('-', '.', rownames(qc_metrics.in))

# Phase 2+3
ids = c('syn26000850', 'syn25976122', 'syn25976136')[-1]
df_read = lapply(ids, function(id){
  qc.in = read.table(synGet(id)$path)
  rownames(qc.in) = gsub('-', '.', rownames(qc.in))
  qc.in
  })

# # sapply(df_read, ncol)
# df_featureCounts_gene = cbind(df_featureCounts_gene, 
#                           df_read[[1]][,-c(1:6)], 
#                           df_read[[2]][,-c(1:6)])

# extract information from sample names
info = data.frame( Name = colnames(geneCounts), stringsAsFactors=FALSE)
rownames(info) = info$Name
info$DissectionID = sapply(strsplit( info$Name, '\\.'), function(x) x[1])
info$CellType = sapply(strsplit( info$Name, '\\.'), function(x) x[2])
info$Assay = sapply(strsplit( info$Name, '\\.'), function(x) x[3])

# set color codes for each cell type
colorCodes = c(
  GABA = '#66A61E',
  GLU = '#E6AB02',
  Olig = '#E7298A',
  MgAs = '#7570B3')
  # GABAGLU_vs_Other = '#A6A910',
  # GABA_vs_GLU = "grey")

# set factor levels to be the same as the order of the colors
# this makes plotting easier
info$CellType = factor(info$CellType,names(colorCodes))

# add information about each sample
info = merge(info, metadata1, by.x="DissectionID", by.y='Dissection.ID', all.x=TRUE)
info = merge(info, qc_metrics, by.x="Name", by.y='row.names', all.x=TRUE)
rownames(info) = info$Name

# info = merge(info, metadataCMC, by.x='Individual.ID', by.y="Individual ID")
#  info[1:4, 1:20]

# Merge with batch info
info = merge(info, df_batch, by="Name")
rownames(info) = info$Name

# set order to be the same as geneCounts
# Must be last change to info
info = info[match(colnames(geneCounts), info$Name),]


# exclude columns with no variance
index = apply(qc_metrics, 2, function(x) var(x, na.rm=TRUE)) > 0
include = which(index==TRUE)
include = include[! names(include) %in% "PF_BASES"]

df = reshape2::melt(info[,c('DissectionID', 'CellType', colnames(qc_metrics)[include])], id.vars=c('DissectionID', 'CellType'))

ggplot(df, aes( x=CellType, y=value, fill=CellType)) + geom_boxplot() + facet_wrap(~variable, scales="free_y", ncol=5) + theme_bw(16) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + scale_fill_manual("Cell Type", values=colorCodes) 
```

```{r excludeIDs}
excludeIDs = c('CMC_MSSM_002')

incl = !colnames(geneCounts) %in% rownames(info[info$Individual.ID %in% excludeIDs,])
geneCounts = geneCounts[,incl]
info = info[! info$Individual.ID %in% excludeIDs,]

identical(rownames(info), colnames(geneCounts))
```


### Study Summary
```{r}
kable(table(info$Dx, info$CellType)) %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

```{r}
kable(table(info$Individual.ID)) %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

### Save processed data
```{r save.processed}
ALL_USED_IDs= unique(c('syn22321012', 'syn22321013', 'syn22321014', 'syn22321016', 'syn2232101', 'syn26000845','syn25976117', 'syn25976127', 'syn22321013', 'syn22321014', 'syn22321016', 'syn26000850', 'syn25976122', 'syn25976136'))

storeData( geneCounts, "geneCounts_merged.RDS", "Merged gene counts for Mol Prof data")
storeData( info, "info_merged.RDS", "Merge metadata for Mol Prof data")
```

# Per cell type analysus
```{r preprocess.CT}

resByCT = lapply(unique(info$CellType), function(CT){

  idx = info$CellType == CT
  isexpr = filterByExpr(geneCounts[,idx])

  dge = DGEList( geneCounts[isexpr,idx] )
  dge = calcNormFactors(dge)

  form = ~ (1|Gender) + (1|Batch)
  vobj = voomWithDreamWeights(dge, form, info[idx,], plot=TRUE, save.plot=TRUE, BPPARAM=SnowParam(12, progressbar=TRUE))

  form = ~ Dx + (1|Gender) + (1|Batch) + Batch + Gender + Age.of.Death
  fit = dream(vobj, form, info[idx,], BPPARAM=SnowParam(12, progressbar=TRUE))
  fit = eBayes(fit)

  resDE = topTable(fit, coef="DxSCZ", number=Inf)
  resids = residuals(fit)

  # save results for this cell type
  storeData( vobj, paste0("voomObj_", CT, ".RDS"), paste0("voom fit for ", CT))
  storeData( fit, paste0("fit_Dx_", CT, ".RDS"), paste0("dream fit for ", CT))
  storeData( resDE, paste0("de_Dx_", CT, ".RDS"), paste0("Differential expression within ", CT))
  storeData( resids, paste0("residuals_", CT, ".RDS"), paste0("Residuals for ", CT))
})
```




# Jooint analysis

```{r preprocess}
# isexpr = rowSums(edgeR::cpm(geneCounts)>1) >= 0.1*ncol(geneCounts)
isexpr = filterByExpr(geneCounts, model.matrix(~CellType, info))
table(isexpr)

dge = DGEList( geneCounts[isexpr,] )
dge = calcNormFactors(dge)
```

```{r TMM}
df = merge(dge$samples, info, by="row.names")

ggplot(df, aes(CellType, norm.factors, fill=CellType)) + geom_boxplot() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + ylab("TMM factor")
```

```{r expression.density}

geneExpr = edgeR::cpm(dge, log=TRUE)

df = data.table(reshape2::melt(geneExpr))
setkey(df, "Var2")
df = merge(df, info[,c("Name", 'CellType')], by.x = "Var2", by.y="Name")

ggplot(df, aes(value, color=CellType)) + geom_density() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlab(bquote(log[2]~CPM))

ggplot(df, aes(value, color=Var2)) + geom_density() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + facet_wrap(~CellType, ncol=4) + xlab(bquote(log[2]~CPM))
```



### PCA
```{r PCA, fig.width=10, fig.height=5}
dcmp = prcomp( t(geneExpr), scale=TRUE)

df_pca = data.frame(dcmp$x)

df_pca = merge(df_pca, info, by.x="row.names", by.y="Name")

varFrac = dcmp$sdev^2 / sum(dcmp$sdev^2)

xlab = paste0('PC1 (', format(varFrac[1]*100, digits=3), '%)')
ylab = paste0('PC2 (', format(varFrac[2]*100, digits=3), '%)')
fig1 = ggplot(df_pca, aes(PC1, PC2, color=CellType, shape=Batch)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlab(xlab) + ylab(ylab)

xlab = paste0('PC2 (', format(varFrac[2]*100, digits=3), '%)')
ylab = paste0('PC3 (', format(varFrac[3]*100, digits=3), '%)')
fig2 = ggplot(df_pca, aes(PC2, PC3, color=CellType, shape=Batch)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlab(xlab) + ylab(ylab)

plot_grid(fig1, fig2 + theme(legend.position="none"), axis="tbr", align="v")
```      


#### Batch corrected
```{r PCA.batch, fig.width=10, fig.height=5}
fit = dream(geneExpr, ~ (1|Batch), info, BPPARAM=SnowParam(12, progressbar=TRUE))
  
dcmp = prcomp( t(residuals(fit)), scale=TRUE)

df_pca = data.frame(dcmp$x)

df_pca = merge(df_pca, info[,c("Name", "CellType", "Batch")], by.x="row.names", by.y="Name")

varFrac = dcmp$sdev^2 / sum(dcmp$sdev^2)

xlab = paste0('PC1 (', format(varFrac[1]*100, digits=3), '%)')
ylab = paste0('PC2 (', format(varFrac[2]*100, digits=3), '%)')
fig1 = ggplot(df_pca, aes(PC1, PC2, color=CellType, shape=Batch)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlab(xlab) + ylab(ylab)

xlab = paste0('PC2 (', format(varFrac[2]*100, digits=3), '%)')
ylab = paste0('PC3 (', format(varFrac[3]*100, digits=3), '%)')
fig2 = ggplot(df_pca, aes(PC2, PC3, color=CellType, shape=Batch)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlab(xlab) + ylab(ylab)

plot_grid(fig1, fig2 + theme(legend.position="none"), axis="tbr", align="v")
```      


#### Correlation between individuals
```{r corr.btw.indivs}
form = ~ (1|CellType) + (1|Batch) + (1|Individual.ID) 
vobj.simple = voomWithDreamWeights(dge, form, info, BPPARAM=SnowParam(12, progressbar=TRUE))

# fit.simple = dream(vobj.simple, form, info, BPPARAM=SnowParam(12, progressbar=TRUE))
# residCompare = residuals(fit.simple) 

residCompare = fitVarPartModel(vobj.simple, form, info, fxn=function(fit){
  library(lme4)  
  fit@resp[['y']] - variancePartition::get_prediction(fit, ~ (1|CellType) + (1|Batch))
}, BPPARAM=SnowParam(12, progressbar=TRUE))

residCompare = do.call(rbind, residCompare)
```


```{r plot_within_btw}
plot_within_btw = function( simMat, design, col=c("lightblue", "lightgreen"),main='',ylim=NULL,...){
  variation = list(type=c(), value=c(), donor1=c(), donor2=c())
  for(i in 1:ncol(design)){
    idx1 = which(design[,i] == 1)   
    idx0 = which(design[,i] == 0)
    C_sub = simMat[idx1, idx1]
    variation$type = append( variation$type, rep("Same Donor", sum(lower.tri(C_sub))) )  
    variation$value = append( variation$value, C_sub[lower.tri(C_sub)])   
    variation$donor1 = append(variation$donor1, rep(colnames(design)[i], length(C_sub[lower.tri(C_sub)]) ))      
    variation$donor2 = append(variation$donor2, rep(colnames(design)[i], length(C_sub[lower.tri(C_sub)]) ))
    variation$type = append( variation$type, rep("Different Donor", length(simMat[idx1,idx0])) )  
    variation$value = append( variation$value, simMat[idx1,idx0])    
    variation$donor1 = append(variation$donor1, rep("", length(simMat[idx1,idx0])) )      
    variation$donor2 = append(variation$donor2, rep("", length(simMat[idx1,idx0])))
  }
  variation = as.data.frame(variation)
  fig = ggplot( variation, aes(type, value)) + geom_violin(scale = "width", aes(fill=type)) + 
    ylab("Variance explained (%)") + xlab("") + geom_boxplot(width = 0.07, 
        fill = "grey", outlier.color = "black") + theme_bw(15) + 
        scale_fill_manual(values = col) + theme(legend.position = "none") + 
        ylab("Correlation between samples") + scale_fill_manual(values=c("grey", "steelblue1"))
    if( main !='' ){
      fig = fig + ggtitle( main )
    }
    if( !is.null(ylim) ){
      fig = fig + ylim( ylim )
    }
    fig = fig + theme(aspect.ratio=1)
    return(list(fig=fig, variation=variation))
}

C = cor( residCompare, method="spearman" )

res = plot_within_btw( C, model.matrix(~Individual.ID+0, info)) 
res$fig
```


```{r predict.celltype}
X = as.matrix(df_pca[,2:200])

df_class = lapply(1:10, function(k){

  fit = hush(multinom(df_pca$CellType ~ X[,1:k,drop=FALSE]))

  data.table(k =k, n_mismatch = sum(predict(fit) != df_pca$CellType))
  })
df_class = do.call(rbind, df_class)

fig1 = ggplot(df_class, aes(factor(k), n_mismatch)) + geom_bar(stat="identity") + theme_classic() + scale_y_continuous(expand=c(0,0)) + ggtitle("Ability to predict each cell type") + theme(aspect.ratio=1)

df_class = lapply(1:10, function(k){

  fit = hush(multinom(df_pca$CellType ~ X[,1:k,drop=FALSE]))

  Y_pred = predict(fit, type="probs")

  df_class = lapply( levels(df_pca$CellType), function(CT){

    y0 = Y_pred[df_pca$CellType == CT,CT]
    y1 = Y_pred[df_pca$CellType != CT,CT]

    pr = pr.curve(scores.class0 = y0, scores.class1 = y1, curve=TRUE, rand.compute=TRUE)

    data.table(k = k, CellType = CT, AUPR = pr$auc.integral, AUPR.rand = pr$rand$auc.integral)
    })
  do.call(rbind, df_class)
  })
df_class = do.call(rbind, df_class)

fig2 = ggplot(df_class, aes(k, AUPR, color=CellType)) + geom_point() + geom_line() + theme_classic() + scale_color_manual(values = colorCodes) + scale_x_continuous(breaks = 1:10) + ggtitle("Ability to predict each cell type") + theme(aspect.ratio=1)

plot_grid(fig1, fig2, axis="tbr", align="v")
```

### Sex check
```{r sex.check}

XIST = 'ENSG00000229807'
UTY = 'ENSG00000183878'

df = data.frame(  Name  = colnames(geneExpr), 
                  XIST  = geneExpr[XIST,],
                  UTY   = geneExpr[UTY,])
df = merge( df, info[,c('Name', 'Gender', 'Individual.ID')], by="Name" )
df$Gender = factor(df$Gender, c("Male", 'Female'))

ggplot(df, aes(XIST, UTY, color=Gender)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=c("blue", "red")) + ggtitle("Sex check")
```

Samples with wrong sex
```{r sex.off}
kable(df[df$Gender == 'Female' & df$UTY > 5,], digits=2, row.names=FALSE) %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)  
```




### Estimate precision weights
```{r voom}
form = ~ (1|CellType) + (1|Individual.ID) + (1|Gender) + (1|Batch)
vobj = voomWithDreamWeights(dge, form, info, plot=TRUE, save.plot=TRUE, BPPARAM=SnowParam(12, progressbar=TRUE))
```

### mvIC
Need to get proper covariates here

```{r mvIC, eval=FALSE}
colnames(info) = gsub(' ', '_', colnames(info))
colnames(info) = gsub(':', '', colnames(info))
colnames(info) = gsub('\\(', '', colnames(info))
colnames(info) = gsub(')', '', colnames(info))
colnames(info) = gsub('/', '', colnames(info))
colnames(info) = gsub('-', '_', colnames(info))
colnames(info) = gsub(',', '', colnames(info))

include = apply(info, 2, function(x) all(!is.na(x)) && var(x, na.rm=TRUE) >0 )

variables = colnames(info)[colnames(info) %in% names(include[include])]

mvSearch = mvForwardStepwise( vobj, ~ CellType, info, variables)
```


### variancePartition
```{r vp}
form = ~ (1|CellType) + (1|Dx) + (1|CellType:Dx) + (1|Individual.ID) + (1|Gender) + (1|Batch)
vp = fitExtractVarPartModel( geneExpr, form, info, BPPARAM=SnowParam(12, progressbar=TRUE))

plotVarPart( sortCols(vp) )
```



## Differential expression analysis
### Cell type
```{r DE_process}
form = ~ 0 + CellType + (1|Individual.ID) + (1|Gender) + (1|Batch)

index = combn(nlevels(info$CellType),2)

lvl = rbind(levels(info$CellType)[index[1,]], levels(info$CellType)[index[2,]])
contrasts = apply(lvl, 2, function(x){
  paste0('CellType', x[1], " - CellType", x[2])
})
names(contrasts) = apply(lvl, 2, function(x){
  paste0(x[1], "_vs_", x[2])
})

contrasts['GABAGLU_vs_Other'] = "(CellTypeGABA + CellTypeGLU)/2 - (CellTypeMgAs + CellTypeOlig)/2"
contrasts['GLU_vs_Other'] = "CellTypeGLU - (CellTypeGABA + CellTypeMgAs + CellTypeOlig)/3"
contrasts['GABA_vs_Other'] = "CellTypeGABA - (CellTypeGLU + CellTypeMgAs + CellTypeOlig)/3"
contrasts['MgAs_vs_Other'] = "CellTypeMgAs - (CellTypeGABA + CellTypeGLU + CellTypeOlig)/3"
contrasts['Olig_vs_Other'] = "CellTypeOlig - (CellTypeGABA + CellTypeGLU + CellTypeMgAs)/3"

# combined contrast matrix
L = makeContrastsDream( form, info, contrasts = contrasts)

fitBasic = dream( vobj, form, info, L=L, BPPARAM=SnowParam(36, progressbar=TRUE))
```

```{r plotContrasts}
plotContrasts(round(L,2)) 
```

```{r DE.analysis}
index = combn(nlevels(info$CellType),2)

df_deCount = lapply( 1:ncol(index), function(j){

    CT1 = levels(info$CellType)[index[1,j]]
    CT2 = levels(info$CellType)[index[2,j]]

    key = paste0(CT1, '_vs_', CT2)

    tabl = topTable(fitBasic, coef=key, number=Inf)

    data.frame(CT1, CT2, key, count = sum(tabl$adj.P.Val < 0.05), pi1 = 1 - qvalue(tabl$P.Value)$pi0)
})
df_deCount = do.call(rbind, df_deCount)

df_deCount$CT1 = factor(df_deCount$CT1, levels(info$CellType))
df_deCount$CT2 = factor(df_deCount$CT2, levels(info$CellType))

ggplot(df_deCount, aes(CT1, CT2, fill=count)) + geom_tile() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + scale_fill_gradient(low="white", high="red", limits=c(0, 30000)) + geom_text(aes(label=format(count, big.mark=','))) + xlab('') + ylab('') + ggtitle("Count differential expression genes")

ggplot(df_deCount, aes(CT1, CT2, fill=pi1)) + geom_tile() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + scale_fill_gradient(name = quote(pi[1]), low="white", high="red", limits=c(0, 1)) + geom_text(aes(label=format(pi1, digits=3))) + xlab('') + ylab('') + ggtitle(bquote(pi[1]))
```

```{r DE2}
index = combn(nlevels(info$CellType),2)

df = lapply( 1:ncol(index), function(j){

    CT1 = levels(info$CellType)[index[1,j]]
    CT2 = levels(info$CellType)[index[2,j]]

    key = paste0(CT1, '_vs_', CT2)

    tabl = topTable(fitBasic, coef=key, number=Inf)

    data.frame(CT1, CT2, key, tabl)
})
df = do.call(rbind, df)

ggplot(df, aes(logFC, ..count..)) + geom_density(alpha = .2) + facet_wrap(~key) + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) 
```

## Plots of top differentially expressed genes
```{r DE.marker_genes, fig.width=15, fig.height=40}
index = combn(nlevels(info$CellType),2)

ensIDs = sapply( 1:ncol(index), function(j){

    CT1 = levels(info$CellType)[index[1,j]]
    CT2 = levels(info$CellType)[index[2,j]]

    key = paste0(CT1, '_vs_', CT2)

    tabl = topTable(fitBasic, coef=key, number=Inf)

    tabl = tabl[order(tabl$t),]

    c(head(rownames(tabl), 10), tail(rownames(tabl), 10))
})

ensIDs = unique(c(ensIDs, 'ENSG00000110693', 'ENSG00000167281'))

df = reshape2::melt(geneExpr[ensIDs,])
df = merge(df, info[,1:3], by.x="Var2", by.y="row.names")

df_hugo = AnnotationDbi::select(org.Hs.eg.db, unique(as.character(df$Var1)), "SYMBOL", "ENSEMBL")
df = merge(df, df_hugo, by.x="Var1", by.y="ENSEMBL")
df = df[!is.na(df$SYMBOL),]

ggplot(df, aes(CellType, value, fill=CellType)) + geom_boxplot() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position = "none", axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + scale_fill_manual("Cell Type", values=colorCodes) + facet_wrap(~SYMBOL, ncol=5) + ylab(bquote(log[2]~CPM)) 
```

# Gene set analysis
```{r compare.celltypes, fig.width=8, fig.height=12}
zenith_gsa = function (fit, coefs, geneSets, n_genes_min = 10, n_genes_max=1000,...){
     
    # from GeneSetCollection to list 
    geneSets = recodeToList(geneSets)
     
    index = ids2indices(geneSets, rownames(fit))
    index = index[sapply(index, length) >= n_genes_min]
    index = index[sapply(index, length) < n_genes_max]

    df_res = lapply(coefs, function(coef) {
        df_res = zenith(fit, coef, index,...)
        data.table(coef = coef, Geneset = rownames(df_res), df_res)
    })
    df_zenith = do.call(rbind, df_res)
    rownames(df_zenith) = c()

    df_zenith$FDR = p.adjust(df_zenith$PValue, "BH")
    df_zenith
}

geneSets_GO = get_GeneOntology("BP")

coefTests = c('GABA_vs_GLU', 'GABAGLU_vs_Other', 'GLU_vs_Other', 'GABA_vs_Other', 'MgAs_vs_Other', 'Olig_vs_Other')
df_zenith = zenith_gsa(fitBasic[fitBasic$Amean > 0,], coefTests, geneSets_GO, inter.gene.cor=0.01)

df_zenith$FDR = p.adjust(df_zenith$p.greater, "BH")

df = df_zenith[,.SD[order(p.greater)[1:3],],by=c("coef")]
cls = unique(df$Geneset)
df = df_zenith[Geneset %in% df$Geneset,]

nrow = length(unique(df$coef))
ncol = length(unique(df$Geneset))

# M = matrix(-log10(df$p.greater), nrow=nrow, ncol=ncol,dimnames=list(unique(df$coef), unique(df$Geneset)), byrow=TRUE)

# M = matrix(-log10(df$p.greater), nrow=nrow, ncol=ncol, byrow=TRUE)
# rownames(M) = unique(df$coef)
# colnames(M) = unique(df$Geneset)

# Heatmap(t(M), col = colorRampPalette(c("white", "red"))(100))

# hcl_col = hclust(dist(M), method="ward.D2")
# df$coef = factor(df$coef, rownames(M)[hcl_col$order])

# hcl_row = hclust(dist(t(M)), method="ward.D2")
# df$Geneset = factor(df$Geneset, colnames(M)[hcl_row$order])
df$Geneset = factor(df$Geneset, cls)
df$label = ifelse(df$FDR < 0.05, '*', '')

ggplot(df, aes(coef, Geneset, fill=-log10(p.greater), label=label)) + geom_tile() + theme_classic() + scale_fill_gradient(limits=c(0, NA), low="white", high="red", name=bquote(-log[10]~P)) + theme(aspect.ratio = ncol/nrow, axis.text.x = element_text(angle = 90, vjust = 1, hjust=1), plot.title = element_text(hjust = 0.5)) + xlab('') + ylab('') + geom_text(vjust=1, hjust=0.5) + ggtitle("Compare cell types")
```




## Analysis using just pairs of samples
```{r pairwise.analysis}

index = combn(nlevels(info$CellType),2)

# for all pairs of cell types
tabList = lapply( 1:ncol(index), function(j){

  CT1 = levels(info$CellType)[index[1,j]]
  CT2 = levels(info$CellType)[index[2,j]]
  key = paste0(CT1, '_vs_', CT2)

  idx = which(info$CellType %in% c(CT1, CT2))

  # isexpr = rowSums(edgeR::cpm(geneCounts[,idx])>1) >= 0.3*ncol(geneCounts[,idx])
  isexpr = filterByExpr(geneCounts[,idx], model.matrix(~CellType, info[idx,]))
  table(isexpr)
  infoSub = droplevels(info[idx,])

  dge = DGEList( geneCounts[isexpr,idx] )
  dge = calcNormFactors(dge)

  form = ~ CellType + Gender + Batch
  vobj = voomWithDreamWeights(dge, form, infoSub, plot=FALSE, save.plot=TRUE, BPPARAM=SnowParam(12, progressbar=TRUE))

  form = ~ CellType + Gender + Batch
  fit = dream( vobj, form, infoSub, BPPARAM=SnowParam(12, progressbar=TRUE))
  fit = eBayes(fit)

  coef = paste0('CellType',levels(infoSub$CellType)[2])

  topTable(fit, coef=coef, number=Inf)
})

names(tabList) = apply(matrix(levels(info$CellType)[index], nrow=2), 2, function(x) paste0(x, collapse='.'))

df_de_summary = lapply( names(tabList), function(key){
  tab = data.table(tabList[[key]])

  data.frame( Comparison  = key, 
              nGenes      = nrow(tab), 
              nDE         = sum(tab$adj.P.Val < 0.05),
              nDE_gt_1    =  nrow(tab[adj.P.Val < 0.05 & abs(logFC) > 1,]), 
              nDE_gt_2    =  nrow(tab[adj.P.Val < 0.05 & abs(logFC) > 2,]), 
              nDE_gt_3    =  nrow(tab[adj.P.Val < 0.05 & abs(logFC) > 3,]), 
              nDE_gt_4    =  nrow(tab[adj.P.Val < 0.05 & abs(logFC) > 4,]), 
              pi1         = 1-qvalue(tab$P.Value)$pi0)
  })
df_de_summary = do.call(rbind, df_de_summary)

kable(df_de_summary) %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

```{r pairwise.density}

index = combn(nlevels(info$CellType),2)

# for all pairs of cell types
tabList = lapply( 1:ncol(index), function(j){

  CT1 = levels(info$CellType)[index[1,j]]
  CT2 = levels(info$CellType)[index[2,j]]
  key = paste0(CT1, '_vs_', CT2)

  idx = which(info$CellType %in% c(CT1, CT2))

  # isexpr = rowSums(edgeR::cpm(geneCounts[,idx])>1) >= 0.3*ncol(geneCounts[,idx])
  isexpr = filterByExpr(geneCounts[,idx], model.matrix(~CellType, info[idx,]))
  table(isexpr)
  infoSub = droplevels(info[idx,])

  dge = DGEList( geneCounts[isexpr,idx] )
  dge = calcNormFactors(dge)

  form = ~ CellType + Gender + Batch
  vobj = voomWithDreamWeights(dge, form, infoSub, plot=FALSE, save.plot=TRUE, BPPARAM=SnowParam(12, progressbar=TRUE))

  form = ~ CellType + Gender + Batch
  fit = dream( vobj, form, infoSub, BPPARAM=SnowParam(12, progressbar=TRUE))
  fit = eBayes(fit)

  coef = paste0('CellType',levels(infoSub$CellType)[2])

  data.frame(key, topTable(fit, coef=coef, number=Inf))
})
df = do.call(rbind, tabList)

ggplot(df, aes(logFC)) + geom_density(alpha = .2) + facet_wrap(~key) + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) 
```


```{r compare.voom}

index = combn(nlevels(info$CellType),2)

# for all pairs of cell types
df_voomTrend = lapply( 1:ncol(index), function(j){

  CT1 = levels(info$CellType)[index[1,j]]
  CT2 = levels(info$CellType)[index[2,j]]
  key = paste0(CT1, '_vs_', CT2)

  idx = which(info$CellType %in% c(CT1, CT2))

  # isexpr = rowSums(edgeR::cpm(geneCounts[,idx])>1) >= 0.3*ncol(geneCounts[,idx])
  isexpr = filterByExpr(geneCounts[,idx], model.matrix(~CellType, info[idx,]))
  table(isexpr)
  infoSub = droplevels(info[idx,])

  dge = DGEList( geneCounts[isexpr,idx] )
  dge = calcNormFactors(dge)

  form = ~ CellType + Gender + Batch
  vobj = voomWithDreamWeights(dge, form, infoSub, plot=FALSE, save.plot=TRUE, BPPARAM=SnowParam(6, progressbar=TRUE))

  data.frame( key, x = vobj$voom.line[['x']], y = vobj$voom.line[['y']])
})
df_voomTrend = do.call(rbind, df_voomTrend)

ggplot(df_voomTrend, aes(x, y, color=key)) + geom_line() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + ylab("sqrt( standard deviation)") + xlab("log2( count size + 0.5 )") + xlim(0, 22) + ylim(0.4, 1.4)
```

## Identify marker genes
```{r tabList.markerGenes}

tabList = lapply(colnames(L), function(coef){
  tab = topTable(fitBasic, coef=coef, number=Inf)
  # tab[which((tab$adj.P.Val < 0.05) & (abs(tab$logFC) > 5)),]
  tab
})
names(tabList) = colnames(L)

saveRDS(tabList, file='./Roussos_MSSM_MolProf_FANS_4_diffExpression.RDS')

# system("ml python; synapse add --parentid syn22243354 Roussos_MSSM_MolProf_FANS_4_diffExpression.RDS")
```




## Plots of top differentially expressed genes
```{r markerGenes, fig.width=15, fig.height=40}

lapply( names(tabList), function(key){tab = tabList[[key]]; nrow(tab[tab$logFC > 6,]) })

boxplot_gene = function(ensIDs, tab, ngenes=5){
  df = reshape2::melt(geneExpr[ensIDs,,drop=FALSE])
  df = merge(df, info[,1:3], by.x="Var2", by.y="row.names")

  df_hugo = AnnotationDbi::select(org.Hs.eg.db, unique(as.character(df$Var1)), "SYMBOL", "ENSEMBL")
  df = merge(df, df_hugo, by.x="Var1", by.y="ENSEMBL")
  df = df[!is.na(df$SYMBOL),]

  df = df[df$SYMBOL %in% df_hugo$SYMBOL[!is.na(df_hugo$SYMBOL)][1:ngenes], ]

  ggplot(df, aes(CellType, value, fill=CellType)) + geom_boxplot() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position = "none") + scale_fill_manual("Cell Type", values=colorCodes) + facet_wrap(~SYMBOL, nrow=1) + ylab(bquote(log[2]~CPM))
}

figList = lapply( names(tabList), function(key){

  tab = tabList[[key]]

  # only select up regulated genes
  tab = tab[tab$logFC > 6,]

  ensIDs = rownames(tab)

  boxplot_gene( ensIDs, tab, 5) + ggtitle(key)
  } )
plot_grid(plotlist = figList, ncol=1)

```

```{r FANS.markers}
ensIDs =  c('ENSG00000110693','ENSG00000167281', 'ENSG00000100146')
boxplot_gene(ensIDs, tab) + ggtitle("FANS markers")
```




# Compare to GTEx
```{r GTEX.voom}

# file_SampleAttributesDS = "/sc/arion/projects/CommonMind/data/resources/GTEx/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt"
file_SampleAttributesDS = synGet('syn22321103')$path

# file_SubjectPhenotypesDS= "/sc/arion/projects/CommonMind/data/resources/GTEx/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt"
file_SubjectPhenotypesDS = synGet('syn22321104')$path

# file_counts = "/sc/arion/projects/CommonMind/data/resources/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct.gz"
file_counts = synGet('syn22321107')$path

gtex_sample = read.delim(file_SampleAttributesDS, header = TRUE, stringsAsFactors = FALSE)
# gtex_sample = gtex_sample[gtex_sample$SMTS == 'Brain',]

rnaseq_sample_id = fread(file_counts, header = TRUE, skip = 2, nrows = 1)
gtex_sample = gtex_sample[gtex_sample$SAMPID %in% colnames(rnaseq_sample_id),]
rownames(gtex_sample) = gtex_sample$SAMPID

# read count
gtex_count <- fread(file_counts, header = TRUE, skip = 2, select = c('Name', unique(gtex_sample$SAMPID)))
gtex_count = as.data.frame(gtex_count, stringsAsFactors=FALSE)

gtex_count$Name = gsub("^(.*)(\\.\\S+)$", "\\1", gtex_count$Name )
gtex_count = gtex_count[!duplicated(gtex_count$Name), ]
rownames(gtex_count) = gtex_count$Name
gtex_count$Name = NULL
gtex_count = as.matrix(gtex_count)

# isexpr = rowSums(cpm(gtex_count>1)) >= 0.3*ncol(gtex_count)
# dge = DGEList( gtex_count[isexpr,] )
# dge = calcNormFactors(dge)
# gtex_log2cpm = cpm(dge, log=TRUE)

# df_voomTrend_GTEx = lapply( unique(gtex_sample$SMTSD), function(Tissue){

#   idx = which( gtex_sample$SMTSD == Tissue)

#   isexpr = rowSums(cpm(gtex_count[,idx])>1) >= 0.3*ncol(gtex_count[,idx])
#   table(isexpr)

#   dge = DGEList( gtex_count[isexpr,idx] )
#   dge = calcNormFactors(dge)

#   vobj = voomWithDreamWeights(dge, ~1, gtex_sample[idx,], plot=FALSE, save.plot=TRUE, BPPARAM=SnowParam(6, progressbar=TRUE))

#   data.frame( key = Tissue, x = vobj$voom.line[['x']], y = vobj$voom.line[['y']])
# })
# df_voomTrend_GTEx = do.call(rbind, df_voomTrend_GTEx)
```

```{r GTEX.voom.plot}
# ggplot(df_voomTrend_GTEx, aes(x, y, color=key)) + geom_line() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="bottom") + ylab("sqrt( standard deviation)") + xlab("log2( count size + 0.5 )") + ggtitle("GTEx brain regions") + xlim(2, 22) + ylim(0.45, 1.36)
```

```{r gtex.colors}
# https://genome.ucsc.edu/cgi-bin/hgTables?db=hg19&hgta_track=hgFixed&hgta_group=allTables&hgta_table=hgFixed.gtexTissue

# file= '/sc/arion/projects/CommonMind/data/resources/GTEx/GTEx_Tissue_colors.csv'
file = synGet('syn22321106')$path
gtexColors = read.csv( file, header=TRUE)
```


```{r CV.all.in.one, eval=TRUE}
library(scran)
# FANS4
# isexpr = rowSums(edgeR::cpm(geneCounts)>1) >= 0.3*ncol(geneCounts)
isexpr = filterByExpr(geneCounts, model.matrix(~CellType, info))

se <- SummarizedExperiment( assays=list(counts = as.matrix(geneCounts[isexpr,])), colData = info)
res = modelGeneCV2( se, block=info$CellType )
fit = metadata(res)

df_fans4 = lapply( names(res$per.block), function(key){
  with( res$per.block[[key]], data.frame(key, mean, total, trend, ratio, FDR))
  })
df_fans4 = data.table(do.call(rbind, df_fans4))


df_fans4 = df_fans4[order(df_fans4$mean),]
ggplot(df_fans4, aes(log2(mean), log2(trend), color=key)) + geom_line() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + xlab(bquote(log[2]~Mean~expression)) + ylab(bquote(log[2]~CV~trend)) + scale_color_manual(values=colorCodes)+ xlim(0, 20) + ylim(-5, 5)


# GTEX
idx = which(gtex_sample$SMTSD %in% names(which(table(gtex_sample$SMTSD) > 30)))

# isexpr = rowSums(edgeR::cpm(gtex_count[,idx])>1) >= 0.3*ncol(gtex_count[,idx])
isexpr = filterByExpr(gtex_count[,idx], model.matrix(~SMTSD, gtex_sample[idx,]))

se <- SummarizedExperiment( assays=list(counts = gtex_count[isexpr,idx]), colData = gtex_sample[idx,])
res = modelGeneCV2( se, block=gtex_sample$SMTSD[idx] )

df_gtex = lapply( names(res$per.block), function(key){
  with( res$per.block[[key]], data.frame(key, mean, total, trend, ratio, FDR))
  })
df_gtex = data.table(do.call(rbind, df_gtex))
df_gtex$key = factor(df_gtex$key)

i = match(levels(df_gtex$key), gtexColors$description)
colsGTEx = as.hexmode(gtexColors$color)[i[!is.na(i)]]


df_gtex = df_gtex[order(df_gtex$mean),]
df_gtex = df_gtex[df_gtex$mean > 0,]

ggplot(df_gtex, aes(log2(mean), log2(trend), color=key)) + geom_line() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + xlab(bquote(log[2]~Mean~expression)) + ylab(bquote(log[2]~CV~trend))+ scale_color_manual(values=colsGTEx) + xlim(0, 20) + ylim(-5, 5)

df_gtex$dataset = 'GTEx'
df_fans4$dataset = "FANS4"
df_combined = rbind(df_gtex, df_fans4)

ggplot(df_combined, aes(log2(mean), log2(trend), color=key, size=dataset)) + geom_line() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + xlab(bquote(log[2]~Mean~expression)) + ylab(bquote(log[2]~CV~trend)) + scale_size_manual(values=c(2, .1)) + scale_color_manual(values=c(rep("black", length(colsGTEx)), colorCodes)) + xlim(-8, 20) + ylim(-5, 3) 
```


```{r coef.variation, eval=TRUE}

CTarray = as.list(levels(info$CellType))
# CTarray[[5]] = c("GABA", "GLU")
# CTarray[[6]] = levels(info$CellType)

# different because only genes expressed in given cell type are used
res_cv_FANS = lapply( unlist(CTarray), function(CT){
   
  id = info$Name[which(info$CellType %in% CT)]

  # isexpr = rowSums(edgeR::cpm(geneCounts[,id])>1) >= 0.1*ncol(geneCounts[,id])
  isexpr = filterByExpr(geneCounts[,id], model.matrix(~CellType, info[id,]))

  se <- SummarizedExperiment( assays=list(counts = as.matrix(geneCounts[isexpr,id])), colData = info[id,])
  res = modelGeneCV2( se ) # returns CV^2
  fit = metadata(res)

  geneTab = AnnotationDbi::select(EnsDb.Hsapiens.v86, keys=names(isexpr[isexpr]), columns=c("GENEID", "GENEBIOTYPE"), keytype="GENEID")
  
  list(df = data.frame( key = paste0(CT, collapse="."), res ),
       f = fit$trend,
       nExpressed = sum(isexpr),
       nExpressed.CDS = sum(geneTab$GENEBIOTYPE == 'protein_coding'))
})

x = 2^seq(-3, 15, length.out=500)
df_trend_FANS = lapply(res_cv_FANS, function(res){ 
  data.frame( Dataset="FANS4", key = res$df$key[1], mean = x, trend=res$f(x))
  })
df_trend_FANS = do.call(rbind, df_trend_FANS)

df_cv_FANS = lapply(res_cv_FANS, function(res){ 
  res$df
  })
df_cv_FANS = do.call(rbind, df_cv_FANS)


colorCodes['GABA.GLU'] = "orange"
colorCodes['GABA.GLU.Olig.MgAs'] = "dodgerblue"

# Take sqrt(CV^2)
ggplot(df_trend_FANS, aes(log2(mean), log2(sqrt(trend)), color=key, size=Dataset)) + geom_line() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position=c(.7, .9)) + xlab(bquote(log[2]~Mean~expression)) + ylab(bquote(log[2]~CV~trend)) + scale_size_manual(values=c(2, .1)) + scale_color_manual(values=colorCodes) + scale_x_continuous(expand=c(0.01, 0))

ggplot(df_cv_FANS[df_cv_FANS$key %in% names(colorCodes)[1:4],], aes(log2(sqrt(total)), color=key)) + geom_density(size=2) + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="right")  + scale_color_manual(values=colorCodes) + xlab(bquote(log[2]~CV))

ggplot(df_cv_FANS, aes(log2(sqrt(total)), color=key)) + geom_density(size=2) + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="right")  + scale_color_manual(values=colorCodes) + xlab(bquote(log[2]~CV))
```

```{r include.gtex}
#  GTEX
res_cv_GTEx = lapply( names(which(table(gtex_sample$SMTSD) > 30)), function(Tissue){

  id = gtex_sample$SAMPID[which( gtex_sample$SMTSD == Tissue)]
  rownames(gtex_sample) = gtex_sample$SAMPID

  # isexpr = rowSums(edgeR::cpm(gtex_count[,id])>1) >= 0.1*ncol(gtex_count[,id])
  isexpr = filterByExpr( gtex_count[,id] )

  se <- SummarizedExperiment( assays=list(counts = gtex_count[isexpr,id]), colData = gtex_sample[id,])
  res = modelGeneCV2( se )
  fit = metadata(res)

  geneTab = AnnotationDbi::select(EnsDb.Hsapiens.v86, keys=names(isexpr[isexpr]), columns=c("GENEID", "GENEBIOTYPE"), keytype="GENEID")  

  list(df = data.frame( key = Tissue, res ),
       f = fit$trend,
       nExpressed = sum(isexpr),
       nExpressed.CDS = sum(geneTab$GENEBIOTYPE == 'protein_coding'))
})

df_trend_GTEx = lapply(res_cv_GTEx, function(res){ 
  data.frame( Dataset="GTEx", key = res$df$key[1], mean = x, trend=res$f(x))
  })
df_trend_GTEx = do.call(rbind, df_trend_GTEx)
df_trend_GTEx$key = factor(df_trend_GTEx$key)

df_cv_GTEx = lapply(res_cv_GTEx, function(res){ 
  res$df
  })
df_cv_GTEx = do.call(rbind, df_cv_GTEx)
df_cv_GTEx$key = factor(df_cv_GTEx$key)


i = match(levels(df_trend_GTEx$key), gtexColors$description)
colsGTEx = as.hexmode(gtexColors$color)[i[!is.na(i)]]

ggplot(df_trend_GTEx, aes(log2(mean), log2(sqrt(trend)), color=key)) + geom_line() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + xlab(bquote(log[2]~Mean~expression)) + ylab(bquote(log[2]~CV~trend)) + scale_color_manual(values=colsGTEx) + scale_x_continuous(expand=c(.01,0))


ggplot(df_cv_GTEx, aes(log2(sqrt(total)), color=key)) + geom_density(size=2) + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + xlab(bquote(log[2]~CV~trend)) + scale_color_manual(values=colsGTEx)



# combine
df_combined = rbind(df_trend_GTEx, df_trend_FANS)

ggplot(df_combined, aes(log2(mean), log2(trend), color=key, size=Dataset)) + geom_line() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + xlab(bquote(log[2]~Mean~expression)) + ylab(bquote(log[2]~CV~trend)) + scale_size_manual(values=c(2, .2)) + scale_color_manual(values=c(rep("grey60", length(colsGTEx)), colorCodes)) + scale_x_continuous(expand=c(.01,0))


df_cv_GTEx$Dataset = "GTEx"
df_cv_GTEx$Type = "Real"
df_cv_FANS$Dataset = "FANS4"
df_cv_FANS$Type = "Real"
df_cv_FANS$Type[df_cv_FANS$key %in% names(colorCodes)[5:6]] = "Synthetic"

df = data.table(rbind(df_cv_GTEx, df_cv_FANS))

ggplot(df, aes(log2(total), color=key, size=Dataset, linetype=Type)) + geom_density(size=2) + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + xlab(bquote(log[2]~CV~trend)) + scale_color_manual(values=c(rep("grey", length(colsGTEx)), colorCodes)) + scale_size_manual(values=c(2, .2)) + scale_linetype_discrete(c("solid", "dashed"))

# number of expressed genes
 df[,sum(total > 0), by='key']


df2 = lapply(sort(unique(df$key)), function(KEY){
  data.frame(key = KEY, Dataset = df$Dataset[df$key == KEY][1],
    Type = df$Type[df$key == KEY][1],
    CV2=df[key==KEY,][order(mean,decreasing=TRUE)[1:15000],]$total)
  })
df2 = data.table(do.call(rbind, df2))
df2$key = factor(df2$key, (sort(unique(df2$key))))
levels(df2$key)

i = match(levels(df2$key), gtexColors$description)
colsGTEx = paste0("#", as.hexmode(gtexColors$color)[i[!is.na(i)]])
names(colsGTEx) = gtexColors$description[i[!is.na(i)]]

cols = c(colsGTEx, colorCodes)[levels(df2$key)]
cols[names(colsGTEx)] = "grey"

ggplot(df2, aes(log2(sqrt(CV2)), color=key, size=Dataset, linetype=Type)) + geom_density(size=2) + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + xlab(bquote(log[2]~CV~trend)) + scale_color_manual(values=cols) + scale_size_manual(values=c(2, .1)) + scale_linetype_discrete(c("solid", "dashed"))


df3 = df2[,median(sqrt(CV2)), by='key']
df3$key = factor(df3$key, rev(levels(df3$key)))

ggplot( df3[!is.na(df3$V1)], aes(key, V1, fill=key)) + geom_bar(stat='identity') + theme_bw(10) + theme(aspect.ratio=5, plot.title = element_text(hjust = 0.5), legend.position="none")  + scale_fill_manual(values=c(colsGTEx, colorCodes)[levels(df3$key)]) + coord_flip() + ylab(bquote(Median~CV)) + scale_y_continuous(expand=c(0,0)) + xlab('') + ylim(0, max(df3$V1, na.rm=TRUE)*1.05)


# Compare to IPS and LCL

# genes expressed

df_count_fans4 = lapply(res_cv_FANS, function(res){  
  data.frame( Dataset="FANS4", key = res$df$key[1], 
    "Expressed genes"=res$nExpressed, "Expressed coding genes"=res$nExpressed.CDS)
 })
df_count_fans4 = do.call(rbind, df_count_fans4) 

df_count_gtex = lapply(res_cv_GTEx, function(res){  
  data.frame( Dataset="GTEx", key = res$df$key[1], 
    "Expressed genes"=res$nExpressed, "Expressed coding genes"=res$nExpressed.CDS)
 })
df_count_gtex = do.call(rbind, df_count_gtex) 

df_count = rbind(df_count_gtex, df_count_fans4)
df_count$key = factor(df_count$key, rev(levels(df2$key)))


i = match(levels(df_count$key), gtexColors$description)
colsGTEx = paste0("#", as.hexmode(gtexColors$color)[i[!is.na(i)]])
names(colsGTEx) = gtexColors$description[i[!is.na(i)]]

df_melt = reshape2::melt(df_count, id.vars=c("Dataset", "key"))

ggplot( df_melt, aes(key, value, fill=key)) + geom_bar(stat='identity') + theme_bw(10) + theme(aspect.ratio=5, plot.title = element_text(hjust = 0.5), legend.position="none")  + scale_fill_manual(values=c(colsGTEx, colorCodes)[levels(df3$key)]) + coord_flip() + ylab("# genes expressed") + scale_y_continuous(expand=c(0,0)) + xlab('') + facet_wrap(~variable)
```

```{r plot.coef.variation, fig.width=12, fig.height=12, eval=FALSE}
# use colors here
# gtexColors and colorCodes

df = rbind(df_cv_GTEx, df_cv_FANS)
df$key = factor(df$key, c(rev(sort(unique(df_cv_FANS$key))), rev(sort(unique(df_cv_GTEx$key)))))

i = match(levels(df$key), gtexColors$description)
cols = as.hexmode(gtexColors$color)[i[!is.na(i)]]
cols = paste0('#', cols)
names(cols) = gtexColors$description[i[!is.na(i)]]

cols = c(cols, colorCodes)[levels(df$key)]

cols[!names(cols) %in% unique(df_cv_FANS$key)] = "black"

# ggplot( df, aes(log2(BCV), color=key)) + geom_density() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="bottom") + facet_wrap(~Dataset, ncol=1) + scale_color_manual("Type", values=cols) 

ggplot(df[,nrow(.SD), by="key"], aes(key, V1, fill=key)) + geom_bar(stat="identity") + theme_bw(10) + theme(aspect.ratio=5, plot.title = element_text(hjust = 0.5), legend.position="none") + scale_fill_manual("Cell Type", values=cols) + coord_flip() + xlab('') + ylab("# genes exprressed")

ggplot( df[CV.ratio>=1,], aes(CV.ratio, ..density.., color=key)) + geom_density(size=2) + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + scale_color_manual("Cell Type", values=cols) + xscale("log2", .format = TRUE) + facet_wrap(~Dataset, ncol=1)

ggplot( df[CV.ratio>=1,], aes(log2(CV.ratio), color=key)) + stat_ecdf() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position="none") + scale_color_manual("Cell Type", values=cols) + facet_wrap(~Dataset, ncol=2) + ylab("eCDF") + scale_x_continuous(expand=c(0,0), limits=c(0, 10)) 


df2 = df[,data.frame(Dataset=Dataset[1], Count = sum(CV.ratio > 2) / length(CV.ratio)),by="key"]

v = df2[Dataset=="GTEx", median(Count)]

ggplot(df2, aes(key, Count, fill=key))  + geom_bar(stat="identity") + theme_bw(10) + theme(aspect.ratio=5, plot.title = element_text(hjust = 0.5), legend.position="none") + scale_fill_manual("Cell Type", values=cols) + coord_flip() + xlab('') + ylab('# genes passing cutoff') + geom_hline(yintercept=v, linetype="dashed", color="grey70")


df2 = df[,data.frame(Dataset=Dataset[1], Count = sum(FDR < .05)/length(FDR)),by="key"]
v = df2[Dataset=="GTEx", median(Count)]



ggplot(df[,sum(FDR < .05),by='key'], aes(key, V1, fill=key))  + geom_bar(stat="identity") + theme_bw(8) + theme(aspect.ratio=5, plot.title = element_text(hjust = 0.5), legend.position="none") + scale_fill_manual("Cell Type", values=cols) + coord_flip() + xlab('') + ylab('# genes passing cutoff') + geom_hline(yintercept=v, linetype="dashed", color="grey70")

```









```{r cv.old}

# cv = function(x){
#   sd(x) / mean(x)
# }


# #   cv_est = apply(geneExpr[,id], 1, cv)

# #   data.frame( key = CT, Gene = names(cv_est), CV = cv_est, stringsAsFactors=FALSE)
# # })
# # df_cv_FANS = do.call(rbind, df_cv_FANS)

# # CV for GTEX
# df_cv_GTEx = lapply( unique(gtex_sample$SMTSD), function(Tissue){

#   id = gtex_sample$SAMPID[which( gtex_sample$SMTSD == Tissue)]

#   cv_est = apply(gtex_log2cpm[,id], 1, cv)

#   data.frame( key = Tissue, Gene = names(cv_est), CV = cv_est, stringsAsFactors=FALSE)
# })
# df_cv_GTEx = do.call(rbind, df_cv_GTEx)

# df = rbind(df_cv_FANS, df_cv_GTEx)


# ggplot( df_cv_FANS, aes(log10(CV), color=key)) + geom_density() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=colorCodes) + xlim(-2, 5) + ylim(0,1.4)

# ggplot( df_cv_GTEx, aes(log10(CV), color=key)) + geom_density() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + xlim(-2, 5) + ylim(0,1.4)


# ggplot( df, aes(log10(CV), color=key)) + geom_density() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + xlim(-2, 5) + ylim(0,1.4)

```

### PCA including GTEx
Computed on residuals

```{r PCA.gtex}
id = gtex_sample$SAMPID[which( gtex_sample$SMTSD == 'Brain - Frontal Cortex (BA9)')]
countsCombined = merge(geneCounts, gtex_count[,id], by="row.names")
rownames(countsCombined) = countsCombined$Row.names
countsCombined$Row.names = NULL

df_info = data.frame(CT = c(as.character(info$CellType), gtex_sample[id,]$SMTSD))

# isexpr = rowSums(edgeR::cpm(countsCombined)>1) >= 0.3*ncol(countsCombined)
isexpr = filterByExpr(countsCombined, model.matrix(~CT, df_info))

dge = DGEList(countsCombined[isexpr,])
dge = calcNormFactors(dge)

geneExprCombined = edgeR::cpm(dge, log=TRUE)

infoExpand = rbind(data.frame(info[,c("Name", "CellType", "Batch")], 
  dataset="FANS4", stringsAsFactors=FALSE), 
  data.frame(Name=colnames(gtex_count[,id]), CellType = "GTEx BA9", dataset="GTEx", Batch = "GTEx", stringsAsFactors=FALSE))
rownames(infoExpand) = infoExpand$Name

# the first PC separates FANS from bulk, so subtract that and use residuals
fit = dream(geneExprCombined, ~ Batch, infoExpand)
resid = residuals(fit, geneExprCombined)

dcmp = prcomp( t(resid), scale=TRUE)

df_pca = data.frame(dcmp$x)

df_pca = merge(df_pca, infoExpand, by.x="row.names", by.y="Name")

varFrac = dcmp$sdev^2 / sum(dcmp$sdev^2)

xlab = paste0('PC1 (', format(varFrac[1]*100, digits=3), '%)')
ylab = paste0('PC2 (', format(varFrac[2]*100, digits=3), '%)')
ggplot(df_pca, aes(PC1, PC2, color=CellType)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=c(colorCodes, 'GTEx BA9' = "grey")) + xlab(xlab) + ylab(ylab)

xlab = paste0('PC2 (', format(varFrac[2]*100, digits=3), '%)')
ylab = paste0('PC3 (', format(varFrac[3]*100, digits=3), '%)')
ggplot(df_pca, aes(PC2, PC3, color=CellType)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=c(colorCodes, 'GTEx BA9' = "grey")) + xlab(xlab) + ylab(ylab)
```

### MDS
computed on log cpm

```{r GTEX.MDS}
d = as.dist(1 - cor(resid, method="spearman")^2)
# d = as.dist(1 - cor(apply(geneExprCombined, 2, RNOmni::rankNorm), method="sp")^2)

df_mds = data.frame(cmdscale( d, k=2))

df_mds = merge(df_mds, infoExpand, by.x="row.names", by.y="Name")

ggplot(df_mds, aes(X1, X2, color=CellType)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=c(colorCodes, 'GTEx BA9' = "grey")) + xlab('Coordinate 1') + ylab('Coordinate 2')

# ggplot(df_mds, aes(X2, X3, color=CellType)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + scale_color_manual("Cell Type", values=c(colorCodes, , 'GTEx BA9' = "grey")) + xlab('Coordinate 2') + ylab('Coordinate 3')
```

## Differential expression with Increasing sample size
```{r subset.analysis}

n_array =  c(3, 5, 10, 20, 28)
sampleRamp = lapply(n_array, function(N){

  df = sapply(levels(info$CellType), function(CT){
    info$Name[which(info$CellType == CT)[1:N]]
    })
  data.frame(df, stringsAsFactors=FALSE)
  })


index = combn(nlevels(info$CellType),2)

tabRampList = lapply( sampleRamp, function(sampleSet){

  infoSub = droplevels(info[info$Name %in% unlist(sampleSet),])
  geneCountSub = geneCounts[,infoSub$Name]

  # for all pairs of cell types
  tabList = lapply( 1:ncol(index), function(j){

    CT1 = levels(infoSub$CellType)[index[1,j]]
    CT2 = levels(infoSub$CellType)[index[2,j]]
    key = paste0(CT1, '.', CT2)

    idx = which(infoSub$CellType %in% c(CT1, CT2))

    message(key, " ", nrow(sampleSet) )
    infoSub2 = droplevels(infoSub[idx,])

    # isexpr = rowSums(edgeR::cpm(geneCountSub[,idx])>1) >= 0.3*ncol(geneCountSub[,idx])
    isexpr = filterByExpr(geneCountSub[,idx], model.matrix(~CellType, info[idx,]))
    table(isexpr)

    dge = DGEList( geneCountSub[isexpr,idx] )
    dge = calcNormFactors(dge)

    form = ~ CellType #+ (1|Batch)
    vobj = voomWithDreamWeights(dge, form, infoSub2, plot=FALSE, save.plot=TRUE, BPPARAM=SnowParam(6, progressbar=TRUE), quiet=TRUE)

    form = ~ CellType #+ Batch
    fit = dream( vobj, form, infoSub2, BPPARAM=SnowParam(6, progressbar=TRUE), quiet=TRUE)
    fit = eBayes(fit)

    coef = paste0('CellType',levels(infoSub2$CellType)[2])

    topTable(fit, coef=coef, number=Inf)
  })
  names(tabList) = apply(matrix(levels(infoSub$CellType)[index], nrow=2), 2, function(x) paste0(x, collapse='.'))
  tabList  
})
```


Just counting coding genes, but analysis was for everything

```{r subset.analysis.plots, fig.width=9}

df_de_summary = lapply( 1:length(tabRampList), function(j){

  tabList = tabRampList[[j]]

  df_de_summary = lapply( names(tabList), function(key){
    tab = tabList[[key]]

    geneTab = AnnotationDbi::select(EnsDb.Hsapiens.v86, keys=rownames(tab), columns=c("GENEID", "GENEBIOTYPE"), keytype="GENEID")

    tab = merge(tab, geneTab, by.x="row.names", by.y="GENEID")

    tab = data.table(tab)
    tab = tab[GENEBIOTYPE=='protein_coding',]

    df_de_summary = data.frame( Comparison  = key, 
                N           = n_array[j],
                nGenes      = nrow(tab), 
                nDE         = sum(tab$adj.P.Val < 0.05),
                nDE_gt_1    =  nrow(tab[adj.P.Val < 0.05 & abs(logFC) > 1,]), 
                nDE_gt_2    =  nrow(tab[adj.P.Val < 0.05 & abs(logFC) > 2,]), 
                nDE_gt_3    =  nrow(tab[adj.P.Val < 0.05 & abs(logFC) > 3,]), 
                nDE_gt_4    =  nrow(tab[adj.P.Val < 0.05 & abs(logFC) > 4,]), 
                pi1         = 1-qvalue(tab$P.Value)$pi0)
    })
   do.call(rbind, df_de_summary)
})
df_de_summary = do.call(rbind, df_de_summary)

ggplot(df_de_summary, aes(N, nDE_gt_1, color=Comparison)) + geom_point() + geom_line() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), legend.position = "right") + ylab("# DE and |logFC| > 1") + xlab("Sample size") + ylim(0,NA)
```











# Disease
```{r DE.disease}
info$I = with(info, paste0(CellType, ".", Dx))

# Disease
form = ~ 0 + I + (1|Individual.ID) + (1|Gender) + (1|Batch)

# contrasts
L = makeContrastsDream(form, info, contrasts=c(
              GABA.Dx = "IGABA.SCZ - IGABA.Control",
              GLU.Dx = "IGLU.SCZ - IGLU.Control",
              MgAs.Dx = "IMgAs.SCZ - IMgAs.Control",
              Olig.Dx = "IOlig.SCZ - IOlig.Control"))

fitDisease = dream( vobj, form, info, L=L, BPPARAM=SnowParam(12, progressbar=TRUE))
```




```{r Disease.analysis.corr}
df_disease_de = lapply(levels(info$CellType), function(CT){

  key = paste0( CT, '.Dx')

  tabl = topTable( fitDisease, coef = key, number=Inf)
  tabl$CellType = CT
  tabl$ensID = rownames(tabl)
  tabl
  })
df_disease_de = data.table(do.call(rbind, df_disease_de))


df_disease_de[,data.frame(count = sum(adj.P.Val < 0.05), pi1 = 1 - qvalue(P.Value)$pi0), by="CellType"] %>% kable(digits=2, row.names=FALSE) %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)

index = combn(nlevels(info$CellType),2)

# for all pairs of cell types
tabList = lapply( 1:ncol(index), function(j){

  CT1 = levels(info$CellType)[index[1,j]]
  CT2 = levels(info$CellType)[index[2,j]]
  
  df_tstat = merge(df_disease_de[CellType==CT1,], df_disease_de[CellType==CT2,], by="ensID")

  r = with(df_tstat, cor(z.std.x, z.std.y, method="spearman"))

  data.frame(CT1, CT2, corr = r)
})
tabList = do.call(rbind, tabList)

tabList %>% kable(digits=2, row.names=FALSE) %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)

tabList$CT1 = factor(tabList$CT1, names(colorCodes)[1:4])
tabList$CT2 = factor(tabList$CT2, names(colorCodes)[1:4])

ggplot(tabList, aes(CT1, CT2, fill=corr)) + geom_tile() + theme_bw(15) + theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + scale_fill_gradient(name = "Correlation", low="white", high="red", limits=c(0, 1)) + geom_text(aes(label=format(corr, digits=3))) + xlab('') + ylab('') + ggtitle("Correlation between t-statistics")

# C = matrix(1, 4,4)
# rownames(C) = levels(info$CellType)
# colnames(C) = levels(info$CellType)
# tabList$CT1 = factor(tabList$CT1, levels(info$CellType))
# tabList$CT2 = factor(tabList$CT2, levels(info$CellType))

# for(i in 1:nrow(tabList)){
#   C[tabList$CT1[i],tabList$CT2[i]] = tabList$corr[i]
#   C[tabList$CT2[i],tabList$CT1[i]] = tabList$corr[i]
# }

# plot(hclust(as.dist(1-C)))
```

```{r disease.upsetr}
de_genes = lapply(levels(info$CellType), function(CT){

  key = paste0( CT, '.Dx')

  tabl = topTable( fitDisease, coef = key, number=Inf)
  tabl$ensID = rownames(tabl)
  rownames(tabl[tabl$adj.P.Val < 0.05,])
  })
names(de_genes) = levels(info$CellType)

lst = fromList(de_genes)
a = upset(lst, keep.order = FALSE, order.by="freq" )

upset(lst, keep.order = FALSE, order.by="freq", sets.bar.color=colorCodes[a$Set_names] )
```



```{r scatterPlots, dev="png"}
library(viridis)

# Compute two dimensional density
get_density <- function(x, y, n = 250) {
  dens <- MASS::kde2d(x = x, y = y, n = n)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}

# make scatter plot with 2D density using viridis colors
plotScatterDensity = function(value1, value2){

  # convert two vectors to a data frame
  df = data.frame(cbind(value1, value2))

  # determine limits of the plot
  lim = with(df, max(abs(c(value1, value2))))
  
  # Compute 2D density
  df$density <- get_density(df$value1, df$value2, n = 100)
  
  # Scatter plot colored by density
  ggplot(df, aes(value1, value2, color=density)) + geom_point(size=.4) + theme_bw(16) + theme(aspect.ratio=1, legend.position="bottom", plot.title = element_text(hjust = 0.5)) + geom_abline(color="red") + geom_vline(xintercept=0, col="grey40", linetype="dashed") + geom_hline(yintercept=0, col="grey40", linetype="dashed") + xlim(-lim, lim) + ylim(-lim, lim) + scale_color_viridis() + geom_smooth(method="lm", se=FALSE, color="darkorange")
}

index = combn(nlevels(info$CellType),2)

# for all pairs of cell types
figList = lapply( 1:ncol(index), function(j){

  CT1 = levels(info$CellType)[index[1,j]]
  CT2 = levels(info$CellType)[index[2,j]]
  
  df_tstat = merge(df_disease_de[CellType==CT1,], df_disease_de[CellType==CT2,], by="ensID")

  fig = with(df_tstat, plotScatterDensity(t.x, t.y))
  fig + theme(legend.position="none") + xlab(CT1) + ylab(CT2)
})

plot_grid(plotlist = figList, ncol=3, nrow=3)
```




```{r mashr}
library(mashr)

df_disease_de = lapply(levels(info$CellType), function(CT){

  key = paste0( CT, '.Dx')

  tab = topTable( fitDisease, coef = key, number=Inf)
  tab$ensID = rownames(tab)
  tab$CellType = CT
  tab$se = with(tab, logFC / t)
  tab
  })
names(df_disease_de) = levels(info$CellType)
df_disease_de = do.call(rbind, df_disease_de)

B.hat = reshape2::dcast( df_disease_de, ensID ~ CellType, value.var='logFC')
rownames(B.hat) = B.hat$ensID
B.hat = as.matrix(B.hat[,-1])

se.hat = reshape2::dcast( df_disease_de, ensID ~ CellType, value.var='se')
rownames(se.hat) = se.hat$ensID
se.hat = as.matrix(se.hat[,-1])

mashObj = mash_set_data( B.hat, se.hat)

m.1by1 = mash_1by1(mashObj)
strong = get_significant_results(m.1by1, 0.05)

U.pca = cov_pca(mashObj,3,subset=strong)
print(names(U.pca))

U.ed = cov_ed(mashObj, U.pca, subset=strong)

U.c = cov_canonical(mashObj)  
resMash = mash(mashObj, c(U.c,U.ed))

plotCorrMatrix(cor(resMash$result$PosteriorMean))
```


```{r upset.mashr}
lst = fromList(apply(resMash$result$lfsr, 2, function(x) names(x[x < 0.05])))

a = upset(lst, keep.order = FALSE, order.by="freq" )

upset(lst, keep.order = FALSE, order.by="freq", sets.bar.color=colorCodes[a$Set_names] )
```

Correlation between cell types
```{r corr.celltype}
C = get_pairwise_sharing( resMash ) 

corrplot( C, addCoef.col = "black", title = "Pairwise sharing",
  col=colorRampPalette(c("blue","white","red"))(200),  cl.lim=c(-1,1) )
```



```{r heterogeneity}
df_het = mclapply( 1:nrow(B.hat), function(j){

  fit = rma( B.hat[j,], sei = se.hat[j,])

  data.frame( EnsID = rownames(B.hat)[j], 
    Isq = fit$I2, 
    Hsq = fit$H2, 
    p.het = fit$QEp)
  },mc.cores = 12)
df_het = data.table(do.call(rbind, df_het))

df_hugo = AnnotationDbi::select(org.Hs.eg.db, df_het$EnsID, "SYMBOL", "ENSEMBL")
df_het = merge(df_het, df_hugo, by.x="EnsID", by.y = "ENSEMBL")
df_het$SYMBOL[is.na(df_het$SYMBOL)] = df_het$EnsID[is.na(df_het$SYMBOL)]

# resMash$result$PosteriorMean[j,]

par(mfrow=c(1,2))
hist(df_het$p.het)
qq(df_het$p.het)

nDE = sum(p.adjust(df_het$p.het, "fdr") < 0.05)
pi1 = 1 - qvalue(df_het$p.het)$pi0

# Gene set analysis of heterogeneity
stat = qchisq(df_het$p.het, 1, lower.tail=FALSE)
names(stat) = df_het$EnsID

geneSets_GO = get_GeneOntology()
geneSets = recodeToList(geneSets_GO)
   
index = ids2indices(geneSets, names(stat))
index = index[sapply(index, length) >= 15]
index = index[sapply(index, length) < 500]

res_het = cameraPR( stat, index, use.ranks=TRUE)

res_het[res_het$FDR < 0.05,] %>% kable() %>% kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

Number of DE genes: `r nDE`. pi1 = `r pi1`


```{r plot.hetero, fig.height=30}

plotEffectSize = function(CellType, beta, se, main=''){

  df = data.frame(CellType, beta, se)

  ggplot(df, aes(CellType, beta, ymin = beta - 1.96*se, ymax = beta + 1.96*se, color=CellType)) + geom_point() + geom_errorbar(width=.1) + theme_classic() + theme(aspect.ratio=1, legend.position="none", plot.title = element_text(hjust = 0.5)) + coord_flip() + ylab(bquote(log[2]~FC)) + scale_color_manual(values = colorCodes[df$CellType]) + geom_hline(yintercept=0, linetype="dashed") + ggtitle(main)
}


figList = lapply( 1:10, function(k){

  j = order(df_het$p.het)[k]
  ensID = df_het$EnsID[j]
  geneID = df_het$SYMBOL[j]

  main = paste0(geneID, ' (', ensID, ")\nHeterogeneity (p=", format(df_het$p.het[j], digits=2), ')')

  fig1 = plotEffectSize( colnames(B.hat),
    B.hat[ensID,],
    se.hat[ensID,],
    main)

  main = paste0(geneID, ' (', ensID, ")\nHeterogeneity (p=", format(df_het$p.het[j], digits=2), ')')

  fig2 = plotEffectSize( colnames(resMash$result$PosteriorMean),
    resMash$result$PosteriorMean[ensID,],
    resMash$result$PosteriorSD[ensID,],
    main)

  plot_grid(fig1, fig2)
})
plot_grid(plotlist = figList, ncol=1)
```






```{r zenith.disease}
plotZenith = function(df){

  nrow = length(unique(df$coef))
  ncol = length(unique(df$Geneset))

  df$Geneset = factor(df$Geneset, cls)
  df$label = ifelse(df$FDR < 0.05, '*', '')

  M = reshape2::dcast(df, coef ~ Geneset, value.var="delta")
  annot = M[,1:2]
  M = as.matrix(M[,-1])
  rownames(M) = annot$coef

  hcl2 <- hclust(dist(t(M)))
  df$Geneset = factor(df$Geneset, colnames(M)[hcl2$order])

  statMax = with(df, max(abs(delta/se)))

  ggplot(df, aes(coef, Geneset, fill=delta/se, label=label)) + geom_tile() + theme_classic() + scale_fill_gradient2(limits=c(-statMax, statMax), low="blue", mid="white", high="red", name="t-statistic") + theme(aspect.ratio = ncol/nrow, axis.text.x = element_text(angle = 90, vjust = 1, hjust=1), plot.title = element_text(hjust = 0.5)) + xlab('') + ylab('') + geom_text(vjust=1, hjust=0.5) + ggtitle("Compare cell types")
}

geneSets_GO = get_GeneOntology()

df_zenith = zenith_gsa(fitDisease[fitDisease$Amean > 0,], colnames(coef(fitDisease))[1:4], geneSets_GO, inter.gene.cor=0.01)

df = df_zenith[,.SD[order(PValue)[1:10],],by=c("coef")]
cls = unique(df$Geneset)
df = df_zenith[Geneset %in% df$Geneset,]

plotZenith(df)
```

```{r zenith.disease.mSigDB, fig.width=12}
geneSets_MSigDB = get_MSigDB(c('H', "C1", "C2", "C3", "C5", "C7"))

df_zenith = zenith_gsa(fitDisease[fitDisease$Amean > 0,], colnames(coef(fitDisease))[1:4], geneSets_MSigDB, inter.gene.cor=0.01)

df = df_zenith[,.SD[order(PValue)[1:10],],by=c("coef")]
cls = unique(df$Geneset)
df = df_zenith[Geneset %in% df$Geneset,]

plotZenith(df)
```





Gene-by-gene correlation
fix PCA/MDS
mvIC
splicing
add bulk, imputed, single cell
zenith


# Integrate CommonMind Consortium
```{r load.cmc}
geneCounts_CMC = readRDS(synGet('syn22416410')$path)
info_CMC = readRDS(synGet('syn22416411')$path)
```



```{r, eval=FALSE}
knitr::knit_exit()
```




Volcano plot colored by cell type

DE from null model
Power vs sample size
Paired versus non-paired
Plot examples of marker genes
distribution of effect size
GTEX brain span.
Compare cell types
Homogeneity of variance for each cell type

For next week: deconvolution of GTEx, PEC, CMC, BrainSpan, BrainVar





bMIND can incorporate the prior of cell type variance/covariance. That is, the K x K covariance matrix for K cell types for each gene. I am wondering if Gabriel can also share this info, if sharing the sample-level data is not possible.

### Save mean expression per cell type
https://www.synapse.org/#!Synapse:syn22231629
```{r save.reference.panel}

# compute mean for each cell type, using precision weights
fitCT = dream( vobj, ~ 0 + CellType, info, BPPARAM=SnowParam(6, progressbar=TRUE))
fitCT = eBayes(fitCT)

df_RefPanel = lapply( colnames(fitCT), function(coef){

  # Extract mean for this cell type
  tabl = topTable(fitCT, coef=coef, number=Inf, sort.by='none')
  tabl$se = tabl$logFC / tabl$t # compute standard error given beta and t

  df = data.frame(tabl[,c('logFC', 'se')])
  rownames(df) = rownames(tabl)
  colnames(df) = paste0(gsub('CellType', '', coef), c('', '.se'))
  df

  })
df_RefPanel = do.call(cbind, df_RefPanel)
df_RefPanel = data.frame(EnsID = rownames(df_RefPanel), df_RefPanel)

exclude = grep("\\.se$", colnames(df_RefPanel))
```


```{r store, eval=FALSE}
# Code
CODE <- Folder(name = "FANS4 processing and testing", parentId = "syn22231629") # Make sure these match the folder name and parent Id in the first chunk.
CODE <- synStore(CODE)


# write mean expression
file = 'Roussos_MSSM_MolProf_FANS_4_log2CPM_reference_panel.csv.gz'
gz1 <- gzfile(file, "w")
write.table(df_RefPanel[,-exclude], gz1, quote=FALSE, sep=",", row.names=FALSE)
close(gz1)

nEXP_OBJ = File(file, 
                name = 'FANS4 reference panel - log2CPM', 
                parentId = CODE$properties$id)
nEXP_OBJ = synStore(nEXP_OBJ, used = Ids, 
                    executed = thisFile)



# write standard error
file = 'Roussos_MSSM_MolProf_FANS_4_log2CPM_se_reference_panel.csv.gz'
gz1 <- gzfile(file, "w")
write.table(df_RefPanel[,c(1,exclude)], gz1, quote=FALSE, sep=",", row.names=FALSE)
close(gz1)

nEXP_OBJ = File(file, 
                name = 'FANS4 reference panel - log2CPM standard error', 
                parentId = CODE$properties$id)
nEXP_OBJ = synStore(nEXP_OBJ, used = Ids, 
                    executed = thisFile)





# full counts
file = 'Roussos_MSSM_MolProf_FANS_4_geneCounts.csv.gz'
gz1 <- gzfile(file, "w")
write.table(data.table(EnsID = rownames(geneCounts), geneCounts), gz1, quote=FALSE, sep=",", row.names=FALSE)
close(gz1)

nEXP_OBJ = File(file, 
                name = 'FANS4 reference panel - full counts', 
                parentId = CODE$properties$id)
nEXP_OBJ = synStore(nEXP_OBJ, used = Ids, 
                    executed = thisFile)


# full counts
file = 'Roussos_MSSM_MolProf_FANS_4_info.csv.gz'
gz1 <- gzfile(file, "w")
write.table(info, gz1, quote=FALSE, sep=",", row.names=FALSE)
close(gz1)

nEXP_OBJ = File(file, 
                name = 'FANS4 reference panel - metadata', 
                parentId = CODE$properties$id)
nEXP_OBJ = synStore(nEXP_OBJ, used = Ids, 
                    executed = thisFile)


# Differential expression
file = "Roussos_MSSM_MolProf_FANS_4_diffExpression.RDS"

nEXP_OBJ = File(file, 
                name = 'FANS4 Differential Expression results', 
                parentId = CODE$properties$id)
nEXP_OBJ = synStore(nEXP_OBJ, used = Ids, 
                    executed = thisFile)


```


### Source
[markdown](`r thisFile`)


