---
title: "Joint analysis of read count data"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---


<!---

cd /hpc/users/hoffmg01/work/covarr-de
# rm -rf molec_profiling_cache/
ml git pandoc
git pull
R

system("git pull")


rmarkdown::render("combined_analysis/combine_all_DLPFC.Rmd")


--->



```{r knit2synapse, eval=FALSE}
library(synapser)
library(knit2synapse)
synLogin()
knit2synapse::createAndKnitToFolderEntity(file = "combined_analysis/combine_all_DLPFC.Rmd",
                                          parentId ="syn22231629",
                                          folderName = 'Joint analysis of DLPFC count data',
                                          overwrite=TRUE, 
                                          knitmd=TRUE)
```


```{r setup, include=FALSE}
suppressPackageStartupMessages({
library(synapser)

synLogin()

library(ggplot2)
library(dplyr)
library(ComplexHeatmap)
library(readxl)
library(gtools)
library(corrplot)
library(pvclust)
library(metafor)
library(data.table)
library(cowplot)
library(kableExtra)
library(githubr)
library(ggrepel)
library(knitr)
library(patchwork)
library(plyr)
library(BiocParallel)

library(limma)
library(edgeR)
library(variancePartition)

})
source("../common_functions/common_functions.R")

knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r read.counts}
# Read raw counts from MPP and HBCC
counts.MPP = fread(synGet('syn21867938')$path, data.table=FALSE)
rownames(counts.MPP) = counts.MPP$Geneid
counts.MPP = counts.MPP[,-c(1:6)]

# HBCC - DLFPC
counts.HBCC = fread(synGet('syn21886235')$path, data.table=FALSE)
rownames(counts.HBCC) = counts.HBCC$Geneid
counts.HBCC = counts.HBCC[,-c(1:6)]

# # HBCC - sgACC
counts.HBCC4 = fread(synGet('syn21867866')$path, data.table=FALSE)
rownames(counts.HBCC4) = counts.HBCC4$Geneid
counts.HBCC4 = counts.HBCC4[,-c(1:6)]

geneCountsMerged = cbind(counts.MPP, counts.HBCC, counts.HBCC4)
```


### Data download
#### Obtain effective count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
downloadFile <- function(id){
  fread(synGet(id)$path, data.table = F)
}
downloadFile_version <- function(id , version){
  fread(synGet(id, version = version)$path, data.table = F)
}

# Get RNASeq QCmetadata
METADATA_QC_DLPFC_ID = 'syn16816488' 
ALL_USED_IDs = METADATA_QC_DLPFC_ID
metadata = downloadFile_version(METADATA_QC_DLPFC_ID, version = 19) %>%
  dplyr::select(`Individual_ID`, Institution, Cohort, `Reported_Gender`, Sex, Ethnicity, ageOfDeath, `PMI_(in_hours)`, Dx, 
                Sample_RNA_ID, one_of('rnaSeq_isolation:RIN',
                                      'rnaSeq_dissection:Brain_Region', 
                                      'rnaSeq_isolation:RNA_Isolation_Batch',
                                      'rnaSeq_report:Ribozero_Batch', 'rnaSeq_report:Library_Batch', 
                                      'rnaSeq_report:Flowcell_Batch', 'rnaSeq_report:Exclude?',
                                      'rnaSeq_report:Mapped_Reads',
                                      'rnaSeq_report:Intragenic_Rate','rnaSeq_report:Intronic_Rate',
                                      'rnaSeq_report:Intergenic_Rate','rnaSeq_report:Genes_Detected',
                                      'rnaSeq_report:Expression_Profiling_Efficiency',
                                      'rnaSeq_report:rRNA_Rate',
                                      'rnaSeq_report:Total_Reads', 'rnaSeq_report:Percent_Aligned','rnaSeq_report:Transcripts_Detected')) 

# Get ancestry vector calculated using gemtools
ANCESTRY_ID = 'syn9922992'
ALL_USED_IDs = c(ALL_USED_IDs, ANCESTRY_ID)
ANCESTRY.HBCC = downloadFile(ANCESTRY_ID) %>% 
  plyr::rename(c(ID = 'SNP_report:Genotyping_Sample_ID'))

ANCESTRY_ID = 'syn2511399'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = ANCESTRY_ID
ANCESTRY.MPP = downloadFile(ANCESTRY_ID) %>% 
  plyr::rename(c('DNA_report..Genotyping.Sample_ID' = 'SNP_report:Genotyping_Sample_ID'))

ANCESTRY = rbind(ANCESTRY.HBCC[,colnames(ANCESTRY.MPP)[-2]], ANCESTRY.MPP[,colnames(ANCESTRY.MPP)[-2]])

# Get genotype ids from synapse. 
GENOTYPE_ID = 'syn16816490'
ALL_USED_IDs = c(ALL_USED_IDs, GENOTYPE_ID)
GENOTYPE = downloadFile(GENOTYPE_ID) %>% 
  dplyr::select(Individual_ID, `SNP_report:Genotyping_Sample_ID`, `SNP_report:Exclude?`) %>% 
  dplyr::inner_join(ANCESTRY) %>% 
  dplyr::filter(is.na(`SNP_report:Exclude?`))
```

### Data preprocessing
```{r preprocess.data}
# Merge metadata
METADATA = metadata %>%
  dplyr::left_join(GENOTYPE) %>%      
  dplyr::rename(Region = `rnaSeq_dissection:Brain_Region`,
                PMI = `PMI_(in_hours)`,
                RIN = `rnaSeq_isolation:RIN`, 
                ReportExclude = `rnaSeq_report:Exclude?`,
                GenotypeExclude = `SNP_report:Exclude?`,
                SampleID = Sample_RNA_ID,
                LibraryBatch = `rnaSeq_report:Library_Batch`,
                FlowcellBatch = `rnaSeq_report:Flowcell_Batch`,
                IsolationBatch = `rnaSeq_isolation:RNA_Isolation_Batch`, #isolation batch to replace ribozero batch
                RibozeroBatch = `rnaSeq_report:Ribozero_Batch`,
                MappedReads = `rnaSeq_report:Mapped_Reads`,
                IntragenicRate = `rnaSeq_report:Intragenic_Rate`, IntronicRate = `rnaSeq_report:Intronic_Rate`, 
                IntergenicRate = `rnaSeq_report:Intergenic_Rate`, GenesDetected = `rnaSeq_report:Genes_Detected`,
                ExpProfEfficiency = `rnaSeq_report:Expression_Profiling_Efficiency`, rRNARate = `rnaSeq_report:rRNA_Rate`,
                TotalReads = `rnaSeq_report:Total_Reads`, AlignmentRate = `rnaSeq_report:Percent_Aligned`, TranscriptsDetected = `rnaSeq_report:Transcripts_Detected`) %>%
  dplyr::select(SampleID, Individual_ID, Institution, Cohort, Reported_Gender,Sex, Ethnicity, ageOfDeath, PMI, Dx, Region,
                RIN, ReportExclude, GenotypeExclude, EV.1, EV.2, EV.3, EV.4, EV.5,
                LibraryBatch, FlowcellBatch, IsolationBatch, RibozeroBatch, MappedReads, TotalReads, GenesDetected, AlignmentRate,
                IntragenicRate, IntergenicRate, IntronicRate, ExpProfEfficiency, rRNARate, TranscriptsDetected) %>%
  		 		 dplyr::filter(SampleID %in% colnames(geneCountsMerged), !is.na(SampleID)) %>%
  		 		 dplyr::filter(Region %in% c('DLPFC')) %>%
			 	 dplyr::mutate(Cohort = forcats::fct_recode(Cohort, `MSSM-Penn-Pitt` = 'MSSM-Penn-Pitt', `NIMH-HBCC`='NIMH-HBCC', `NIMH-HBCC`='NIMH-HBCC-4')) %>%
			 	 dplyr::filter(Dx %in% c('Control', "SCZ")) 


# Set Control to baseline
METADATA$Dx = factor(METADATA$Dx, c('Control', 'SCZ'))

ind = METADATA$SampleID[which(METADATA$ReportExclude == 1 | METADATA$GenotypeExclude)]
writeLines(paste('Following',length(ind),'samples are marked exclude'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$Ethnicity) | is.na(METADATA$Institution) | is.na(METADATA$Dx)]
writeLines(paste('Following', length(ind), 'counts are missing any metadata'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$PMI)]
writeLines(paste('Following', length(ind), 'counts are missing PMI'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$Reported_Gender)]
writeLines(paste('Following', length(ind), 'counts are missing gender'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$ageOfDeath)]
writeLines(paste('Following', length(ind), 'counts are missing age of death'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind))

ind = METADATA$SampleID [is.na(METADATA$EV.1)]
writeLines(paste('Following', length(ind), 'counts are missing ancestry information'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind))
```

```{r preprocess.data1, results='asis'}
# Match covariates to expression data
indToRetain = intersect(METADATA$SampleID, colnames(geneCountsMerged))
  
geneCountsMerged = geneCountsMerged[,indToRetain]
 
rownames(METADATA) = METADATA$SampleID
METADATA = METADATA[indToRetain,]
          
# METADATA %>% 
#   group_by(Dx) %>%  
#   dplyr::summarise(count = n()) %>% 
#   kable() 
```

## Summarize samples
```{r summary}
METADATA %>% group_by(Dx, Institution) %>% dplyr::summarise(count = n()) %>% 
  kable(caption="Diagnosis, Institution") %>% kable_styling(full_width=FALSE) 

METADATA %>% group_by( Institution, Reported_Gender) %>% dplyr::summarise(count = n()) %>% 
  kable(caption="Institution, Sex") %>% kable_styling(full_width=FALSE) 

METADATA %>% group_by(Dx, Institution, Reported_Gender) %>% dplyr::summarise(count = n()) %>% 
  kable(caption="Diagnosis, Institution, Sex") %>% kable_styling(full_width=FALSE) 
```


```{r process.counts}
# keep genes that are express in at least one combination of cohort and gender
isexpr.lst = dlply(METADATA, .(Cohort, Reported_Gender), .fun = function(mtd){
	rowSums(cpm(geneCountsMerged[,mtd$SampleID])>1) >= 0.5*ncol(geneCountsMerged[,mtd$SampleID])
	})

isexpr = do.call(rbind, isexpr.lst)
isexpr = colSums(isexpr) >= 3
table(isexpr)
   
# Normalize each cohort separately
dge.lst = lapply( levels(METADATA$Cohort), function(Cohort){
  # normalize counts
  dge = DGEList( geneCountsMerged[isexpr,METADATA$Cohort==Cohort] )
  calcNormFactors(dge) # TMM
})
names(dge.lst) = levels(METADATA$Cohort)


# form = ~ (1|Dx) + Reported_Gender + RIN + ageOfDeath + scale(IntronicRate)*Cohort + scale(IntragenicRate)*Cohort + scale(IntergenicRate)*Cohort + scale(rRNARate)*Cohort + (1|Institution:Cohort) + (1|Region:Cohort) + (1|Individual_ID)

form = ~ Dx + Reported_Gender + RIN + ageOfDeath + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate)

vobj.lst = lapply(dge.lst, function(dge){

  i = match(colnames(dge), rownames(METADATA))
  info = METADATA[i,]

  if( length(unique(info$Institution)) > 1 ){
    form_use = as.formula(paste('~',as.character(form)[2], "+ Institution", collapse=" "))
  }else{
    form_use = form
  }

  voomWithDreamWeights(dge, form_use, info, save.plot=TRUE, BPPARAM = SnowParam(18))
  })
names(vobj.lst) = names(dge.lst)


# FORMULAs

# HBCC - DLPFC: 
# "Dx + scale(RIN) + ageOfDeath + (1 | Reported_Gender) + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate)"

# HBCC dsACC: 
# Dx + scale(RIN) + (1 | Reported_Gender) + ageOfDeath + (1 | IsolationBatch) + scale(IntragenicRate) + scale(IntronicRate) + scale(IntergenicRate)"

# MPP ACC
 # Dx + scale(RIN) + scale(ageOfDeath) + (1 | Institution) + (1 | Reported_Gender) + scale(RIN2) + scale(IntronicRate) + scale(rRNARate) + scale(IntragenicRate) + scale(IntergenicRate)

# MPP DLPFC
# Dx + scale(RIN2) + (1 | Institution) + scale(ageOfDeath) + scale(RIN) + (1 | Reported_Gender) + scale(IntronicRate) + scale(IntragenicRate)
```

### Voom trend plots
```{r plot.voom, fig.height=5, fig.width=10}
fig1 = plot_voom(vobj.lst[[1]]) + ggtitle(names(vobj.lst)[1])
fig2 = plot_voom(vobj.lst[[2]]) + ggtitle(names(vobj.lst)[2])
plot_grid( fig1, fig2, nrow=1) 
```


### Create residuals
```{r Create residuals}

form = ~ Dx*Reported_Gender + RIN + ageOfDeath + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate)

residuals.lst = lapply( vobj.lst, function(vobj){

  i = match(colnames(vobj), rownames(METADATA))
  info = METADATA[i,]

  if( length(unique(info$Institution)) > 1 ){
    form_use = as.formula(paste('~',as.character(form)[2], "+ Institution", collapse=" "))
  }else{
    form_use = form
  }

  # Compute residuals which adding back Diagnosis and Sex effects
  resid.lst = fitVarPartModel(vobj, form, info, fxn = function(fit){
    residuals(fit) + get_prediction( fit, ~ Dx*Reported_Gender)
    }, BPPARAM = SnowParam(18))  
  do.call(rbind, resid.lst)
  })
names(residuals.lst) = names(vobj.lst)
```


### Test marginals
```{r differential.expression.simple}

form = ~ Dx + Reported_Gender + RIN + ageOfDeath + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate)

fit_marginal.lst = lapply( vobj.lst, function(vobj){

  i = match(colnames(vobj), rownames(METADATA))
  info = METADATA[i,]

  if( length(unique(info$Institution)) > 1 ){
    form_use = as.formula(paste('~',as.character(form)[2], "+ Institution", collapse=" "))
  }else{
    form_use = form
  }

  fit = dream(vobj, form, info, BPPARAM = SnowParam(18))  
  eBayes( fit )
  })
names(fit_marginal.lst) = names(vobj.lst)
```


```{r load.genesets}
library(zenith)

# Load Gene Ontology database 
gs.GO = get_GeneOntology()
gs.MSigDB = get_MSigDB()

# convert from GeneSetCollection to list used by camera and zenith
gs.GO.collection = recodeToList( gs.GO )
gs.MSigDB.collection = recodeToList( gs.MSigDB )
```

```{r run.zenith}

coefTest = c('DxSCZ', 'Reported_GenderMale')
res_zenith = lapply( names(fit_marginal.lst), function(key){

  fit = fit_marginal.lst[[key]]

  res.GO = run_zenith( fit, coefTest, gs.GO.collection )
  res.GO$Resource = "GO"
  res.GO$Cohort = key

  res.MSigDB = run_zenith( fit, coefTest, gs.MSigDB.collection )
  res.MSigDB$Resource = "MSigDB"
  res.MSigDB$Cohort = key

  rbind(res.GO, res.MSigDB)

})
names(res_zenith) = names(fit_marginal.lst)
```

```{r get.signatures, fig.width=10, fig.height=5}
# Get DE signatures
df_DE_results = lapply(fit_marginal.lst, function(fit){
  res = lapply(coefTest, function(coef){
    res = topTable(fit, coef, number=Inf) 
    res$Coef = coef
    res
    })
  do.call(rbind, res) %>% getGeneSymbol
  })
names(df_DE_results) = names(fit_marginal.lst)
```

```{r test.scz, fig.height=5, fig.width=10}
# Schizophrenia testing
# extract from each chort
coef = 'DxSCZ'
figList.Scz = lapply( names(df_DE_results), function(key){
  df = df_DE_results[[key]]
  df = df[df$Coef == coef,]

  main = paste(key, coef, sep = ': ' )
  plotVolcano( df, df$Symbol[1:20] ) + ggtitle( main )
})

# volcano plots
plot_grid(plotlist=figList.Scz, nrow=1)
```

```{r meta.analysis.SCZ}
# Create list preparing for meta-analysis for SCZ
tab.lst = lapply( df_DE_results, function(df){
  df[df$Coef == coef,]
})
names(tab.lst) = names(df_DE_results)

# Run meta-analysis
df_meta_SCZ = run_meta_analysis( tab.lst, "FE") %>% getGeneSymbol
```

```{r concordance.scz, fig.height=6, fig.width=10}
# Concordance analysis
resConcord.SCZ = plot_concordance(tab.lst, showGenes = df_meta_SCZ$Symbol[1:20])

main = "Concordance of SCZ"
plot_grid(  resConcord.SCZ$fig_logFC + ggtitle(main),
            resConcord.SCZ$fig_tstat + ggtitle(main), nrow=1)

data.frame(pi1 = resConcord.SCZ$pi_single) %>% kable(caption="pi1 within dataset") %>% kable_styling(full_width=FALSE)
resConcord.SCZ$tab_pi %>% kable(caption="pi1 for replication") %>% kable_styling(full_width=FALSE)
resConcord.SCZ$tab_corr %>% kable(caption="Correlation between t-statistics") %>% kable_styling(full_width=FALSE)
```

```{r plot.meta.analysis.SCZ, fig.height=5, fig.width=10}
# volcano plot for meta-analysis
main = paste('Meta-analysis', coef, sep = ': ' )
plotVolcano( df_meta_SCZ, df_meta_SCZ$Symbol[1:20] ) + ggtitle( main )
```









```{r exit2, cache=FALSE}
knitr::knit_exit()
```

#   rmarkdown::render("combined_analysis/combine_all_DLPFC.Rmd")


```{r show.gene_set_test.marginal}

show_zenith_results = function( resList, key, FDR = 0.05){
	df = resList[[key]]

	# filter by FDR
	df = df[df$FDR <= FDR,]

	df %>% kable(caption=key) %>% kable_styling(full_width=FALSE)
}

show_zenith_results( res_zenith_marginal.GO, "Scz")
show_zenith_results( res_zenith_marginal.MSigDB, "Scz")

show_zenith_results( res_zenith_marginal.GO, "BP")
show_zenith_results( res_zenith_marginal.MSigDB, "BP")

show_zenith_results( res_zenith_marginal.GO, "Sex")
show_zenith_results( res_zenith_marginal.MSigDB, "Sex")
```

# Create residuals using fitVarPart and get_prediction()






### Test Diagnosis * Sex interaction
```{r differential.expression.interaction.DxSex}

# METADATA$Reported_Gender = factor(METADATA$Reported_Gender)
# METADATA$Dx = factor(METADATA$Dx, c("BP", "Control", "SCZ")) 

form = ~ Dx*Reported_Gender + RIN + ageOfDeath + scale(IntronicRate)*Cohort + scale(IntragenicRate)*Cohort + scale(IntergenicRate)*Cohort + scale(rRNARate)*Cohort + (1|Institution:Cohort) + (1|Region:Cohort) + (1|Individual_ID)

# get contrasts
Lbase = getContrast( vobj, form, METADATA, "RIN")
Lbase[] = 0

names(Lbase)

L_SCZ = Lbase
L_SCZ['DxSCZ:Reported_GenderMale'] = 1

L_BP = Lbase
L_BP['DxBP:Reported_GenderMale'] = 1

L = cbind(	SczSex = L_SCZ,
			BPSex = L_BP)
			
plotContrasts(L) + theme(axis.text.x=element_text(angle = 45, vjust = 1, hjust=1))

fit_interaction_DxSex = dream(vobj, form, METADATA, L=L, BPPARAM = SnowParam(18))
```


```{r plots.interaction.DxSex, fig.width=10, fig.height=5}
# Disease. 
tab = topTable(fit_interaction_DxSex, "SczSex", number=Inf) %>% getGeneSymbol
fig1 = plotVolcano( tab, tab$Symbol[1:20] ) + ggtitle("SCZ")
   
tab = topTable(fit_interaction_DxSex, "BPSex", number=Inf)  %>% getGeneSymbol
fig2 = plotVolcano( tab, tab$Symbol[1:20] ) + ggtitle("BP")

plot_grid( fig1, fig2, ncol=2) 
```


```{r gene_set_test.DxSex}
res_zenith_DxSex.GO = run_zenith( fit_interaction_DxSex, colnames(L), gs.GO.collection )
res_zenith_DxSex.MSigDB = run_zenith( fit_interaction_DxSex, colnames(L), gs.MSigDB.collection )     
```

```{r show.gene_set_test.DxSex}
show_zenith_results( res_zenith_DxSex.GO, "DxSCZ:Reported_GenderMale")
show_zenith_results( res_zenith_DxSex.MSigDB, "DxSCZ:Reported_GenderMale")
      
show_zenith_results( res_zenith_DxSex.GO, "DxBP:Reported_GenderMale")
show_zenith_results( res_zenith_DxSex.MSigDB, "DxBP:Reported_GenderMale")
```



```{r exit, cache=FALSE}
knitr::knit_exit()
```



### Test Region * Sex interaction
```{r differential.expression.interaction.RegionSex, eval=FALSE}

form = ~ 0 + Dx + Region*Reported_Gender + RIN + ageOfDeath + scale(IntronicRate)*Cohort + scale(IntragenicRate)*Cohort + scale(IntergenicRate)*Cohort + scale(rRNARate)*Cohort + (1|Institution:Cohort) + (1|Cohort) + (1|Individual_ID)

# get contrasts
Lbase = getContrast( vobj, form, METADATA, "RIN")
Lbase[] = 0

# SCZ by sex
L_SCZ = Lbase
L_SCZ['DxSCZ:Reported_GenderMale'] = 1
L_SCZ['DxControl:Reported_GenderMale'] = -1
L_SCZ['DxSCZ'] = -1
L_SCZ['DxControl'] = 1 
# (A - B) - (C - D) = A - B - C + D

L = cbind(	SczSex = L_SCZ)
			
plotContrasts(L) + theme(axis.text.x=element_text(angle = 45, vjust = 1, hjust=1))

fit_interaction_RegionSex = dream(vobj, form, METADATA, L=L,BPPARAM = SnowParam(18))
```



### Test Region * Disease
```{r differential.expression.interaction.RegionDx, eval=FALSE}

form = ~ 0 + Region*Dx + Reported_Gender + RIN + ageOfDeath + scale(IntronicRate)*Cohort + scale(IntragenicRate)*Cohort + scale(IntergenicRate)*Cohort + scale(rRNARate)*Cohort + (1|Institution:Cohort) + (1|Cohort) + (1|Individual_ID)

# get contrasts
Lbase = getContrast( vobj, form, METADATA, "RIN")
Lbase[] = 0

# SCZ by sex
L_SCZ = Lbase
L_SCZ['DxSCZ:Reported_GenderMale'] = 1
L_SCZ['DxControl:Reported_GenderMale'] = -1
L_SCZ['DxSCZ'] = -1
L_SCZ['DxControl'] = 1 
# (A - B) - (C - D) = A - B - C + D

L = cbind(	SczSex = L_SCZ)
			
plotContrasts(L) + theme(axis.text.x=element_text(angle = 45, vjust = 1, hjust=1))

fit_interaction_RegionSex = dream(vobj, form, METADATA, L=L,BPPARAM = SnowParam(18))
```

































