---
title: "Evaluate PEER in each cohort and brain region"
subtitle: ''
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.time()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---


<!---

cd /hpc/users/hoffmg01/work/CMC/covarr-de
# rm -rf combined_analysis/eval_resid_in_each_cache/
export R_LIBS=/hpc/users/hoffmg01/.Rlib/R_403
ml purge
ml git pandoc R/3.6.3 # to get peer
git pull
R

system("git pull")


rmarkdown::render("combined_analysis/eval_resid_in_each.Rmd")



--->

```{r knit2synapse, eval=FALSE}
library(synapser)
library(knit2synapse)
synLogin()
knit2synapse::createAndKnitToFolderEntity(file = "combined_analysis/eval_resid_in_each.Rmd",
                                          parentId ="syn22012942",
                                          folderName = 'Evaluate residuals in each cohort and brain region',
                                          overwrite=TRUE, 
                                          knitmd=TRUE)
```


```{r setup, include=FALSE}
suppressPackageStartupMessages({
library(synapser)

synLogin()

library(ggplot2)
library(dplyr)
library(ComplexHeatmap)
library(readxl)
library(gtools)
library(corrplot)
library(pvclust)
library(metafor)
library(data.table)
library(peer)
library(cowplot)
library(kableExtra)
library(githubr)
library(ggrepel)
library(knitr)
library(patchwork)
library(plyr)
library(compositions)
library(BiocParallel)

library(limma)
library(edgeR)
library(variancePartition)

})
source("../common_functions/common_functions.R")

knitr::opts_chunk$set(
  echo = FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```




```{r synapse.parameters, include=FALSE, cache=TRUE}

library(githubr)
parentId = 'syn22012942';
activityName = 'Evaluate PEER in each cohort and brain region';
activityDescription = 'Evaluate PEER in each cohort and brain region';

thisFileName <- 'eval_resid_in_each.Rmd'

# Github link
thisRepo <- getRepo(repository = "CommonMindConsortium/covarr-de", ref="branch", refName='master')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('combined_analysis/',thisFileName))
```






```{r read.counts}
# Read raw counts from MPP and HBCC
counts.MPP = fread(synGet('syn21867938')$path, data.table=FALSE)
rownames(counts.MPP) = counts.MPP$Geneid
counts.MPP = counts.MPP[,-c(1:6)]

# HBCC - DLFPC
counts.HBCC = fread(synGet('syn21886235')$path, data.table=FALSE)
rownames(counts.HBCC) = counts.HBCC$Geneid
counts.HBCC = counts.HBCC[,-c(1:6)]

# HBCC - sgACC
counts.HBCC4 = fread(synGet('syn21867866')$path, data.table=FALSE)
rownames(counts.HBCC4) = counts.HBCC4$Geneid
counts.HBCC4 = counts.HBCC4[,-c(1:6)]

# NIMH-HBCC-5
counts.HBCC5 = fread(synGet('syn21867762')$path, data.table=FALSE)
rownames(counts.HBCC5) = counts.HBCC5$Geneid
counts.HBCC5 = counts.HBCC5[,-c(1:6)]

geneCountsMerged = cbind(counts.MPP, counts.HBCC, counts.HBCC4, counts.HBCC5)
```


### Data download
#### Obtain effective count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
downloadFile <- function(id){
  fread(synGet(id)$path, data.table = F)
}
downloadFile_version <- function(id , version){
  fread(synGet(id, version = version)$path, data.table = F)
}

# Get RNASeq QCmetadata
METADATA_QC_DLPFC_ID = 'syn16816488' 
ALL_USED_IDs = METADATA_QC_DLPFC_ID
metadata = downloadFile_version(METADATA_QC_DLPFC_ID, version = 19) %>%
  dplyr::select(`Individual_ID`, Institution, Cohort, `Reported_Gender`, Sex, Ethnicity, ageOfDeath, `PMI_(in_hours)`, Dx, 
                Sample_RNA_ID, one_of('rnaSeq_isolation:RIN',
                                      'rnaSeq_dissection:Brain_Region', 
                                      'rnaSeq_isolation:RNA_Isolation_Batch',
                                      'rnaSeq_report:Ribozero_Batch', 'rnaSeq_report:Library_Batch', 
                                      'rnaSeq_report:Flowcell_Batch', 'rnaSeq_report:Exclude?',
                                      'rnaSeq_report:Mapped_Reads',
                                      'rnaSeq_report:Intragenic_Rate','rnaSeq_report:Intronic_Rate',
                                      'rnaSeq_report:Intergenic_Rate','rnaSeq_report:Genes_Detected',
                                      'rnaSeq_report:Expression_Profiling_Efficiency',
                                      'rnaSeq_report:rRNA_Rate',
                                      'rnaSeq_report:Total_Reads', 'rnaSeq_report:Percent_Aligned','rnaSeq_report:Transcripts_Detected')) 

# Ethnicity
# Genotype_Inferred_Ancestry
# Race

# Get ancestry vector calculated using gemtools
ANCESTRY_ID = 'syn9922992'
ALL_USED_IDs = c(ALL_USED_IDs, ANCESTRY_ID)
ANCESTRY.HBCC = downloadFile(ANCESTRY_ID) %>% 
  plyr::rename(c(ID = 'SNP_report:Genotyping_Sample_ID'))

ANCESTRY_ID = 'syn2511399'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = ANCESTRY_ID
ANCESTRY.MPP = downloadFile(ANCESTRY_ID) %>% 
  plyr::rename(c('DNA_report..Genotyping.Sample_ID' = 'SNP_report:Genotyping_Sample_ID'))

ANCESTRY = rbind(ANCESTRY.HBCC[,colnames(ANCESTRY.MPP)[-2]], ANCESTRY.MPP[,colnames(ANCESTRY.MPP)[-2]])

# Get genotype ids from synapse. 
GENOTYPE_ID = 'syn16816490'
ALL_USED_IDs = c(ALL_USED_IDs, GENOTYPE_ID)
GENOTYPE = downloadFile(GENOTYPE_ID) %>% 
  dplyr::select(Individual_ID, `SNP_report:Genotyping_Sample_ID`, `SNP_report:Exclude?`) %>% 
  dplyr::inner_join(ANCESTRY) %>% 
  dplyr::filter(is.na(`SNP_report:Exclude?`))
```

```{r ancestry, fig.height=6}
ggplot(ANCESTRY.MPP, aes(EV.1, EV.2, color=Cluster)) + geom_point() + ggtitle("MPP ancestry")  + theme(aspect.ratio=1)
ggplot(ANCESTRY.HBCC, aes(EV.1, EV.2, color=as.character(cluster))) + geom_point() + ggtitle("HBCC ancestry")  + theme(aspect.ratio=1)
ggplot(ANCESTRY.HBCC, aes(EV.1, EV.2, color=Population)) + geom_point() + ggtitle("HBCC ancestry") + theme(aspect.ratio=1)
```



### Data preprocessing
```{r preprocess.data}
# Merge metadata
METADATA = metadata %>%
  dplyr::left_join(GENOTYPE) %>%      
  dplyr::rename(Region = `rnaSeq_dissection:Brain_Region`,
                PMI = `PMI_(in_hours)`,
                RIN = `rnaSeq_isolation:RIN`, 
                ReportExclude = `rnaSeq_report:Exclude?`,
                GenotypeExclude = `SNP_report:Exclude?`,
                SampleID = Sample_RNA_ID,
                LibraryBatch = `rnaSeq_report:Library_Batch`,
                FlowcellBatch = `rnaSeq_report:Flowcell_Batch`,
                IsolationBatch = `rnaSeq_isolation:RNA_Isolation_Batch`, #isolation batch to replace ribozero batch
                RibozeroBatch = `rnaSeq_report:Ribozero_Batch`,
                MappedReads = `rnaSeq_report:Mapped_Reads`,
                IntragenicRate = `rnaSeq_report:Intragenic_Rate`, IntronicRate = `rnaSeq_report:Intronic_Rate`, 
                IntergenicRate = `rnaSeq_report:Intergenic_Rate`, GenesDetected = `rnaSeq_report:Genes_Detected`,
                ExpProfEfficiency = `rnaSeq_report:Expression_Profiling_Efficiency`, rRNARate = `rnaSeq_report:rRNA_Rate`,
                TotalReads = `rnaSeq_report:Total_Reads`, AlignmentRate = `rnaSeq_report:Percent_Aligned`, TranscriptsDetected = `rnaSeq_report:Transcripts_Detected`) %>%
  dplyr::select(SampleID, Individual_ID, Institution, Cohort, Reported_Gender,Sex, Ethnicity, ageOfDeath, PMI, Dx, Region,
                RIN, ReportExclude, GenotypeExclude, EV.1, EV.2, EV.3, EV.4, EV.5,
                LibraryBatch, FlowcellBatch, IsolationBatch, RibozeroBatch, MappedReads, TotalReads, GenesDetected, AlignmentRate,
                IntragenicRate, IntergenicRate, IntronicRate, ExpProfEfficiency, rRNARate, TranscriptsDetected) %>%
  dplyr::filter(SampleID %in% colnames(geneCountsMerged), !is.na(SampleID)) %>%
  dplyr::mutate(Cohort = forcats::fct_recode(Cohort, `MSSM-Penn-Pitt` = 'MSSM-Penn-Pitt', `NIMH-HBCC`='NIMH-HBCC', `NIMH-HBCC`='NIMH-HBCC-4')) %>%
  dplyr::filter(ageOfDeath > 17) 

# Set Control to baseline
METADATA$Dx = factor(METADATA$Dx)#, c('Control', 'SCZ'))

ind = METADATA$SampleID[which(METADATA$ReportExclude == 1 | METADATA$GenotypeExclude)]
writeLines(paste('Following',length(ind),'samples are marked exclude'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$Ethnicity) | is.na(METADATA$Institution) | is.na(METADATA$Dx)]
writeLines(paste('Following', length(ind), 'counts are missing any metadata'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$PMI)]
writeLines(paste('Following', length(ind), 'counts are missing PMI'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$Reported_Gender)]
writeLines(paste('Following', length(ind), 'counts are missing gender'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$ageOfDeath)]
writeLines(paste('Following', length(ind), 'counts are missing age of death'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind))

ind = METADATA$SampleID [is.na(METADATA$EV.1)]
writeLines(paste('Following', length(ind), 'counts are missing ancestry information'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind))

# When individuals were sequenced multiple times from the same brain region
#   for technical replicates or QC failure,
# Keep sample with most reads
METADATA = METADATA %>%
  group_by(Individual_ID, Region, Cohort) %>% 
  slice_max(MappedReads)
```


```{r cellFractionEstimates}
# include estimated cell fractions in METADATA
df_cellFractions = read.table(synGet('syn22333694')$path, row.names=1)

celFrac_ilr = ilr(df_cellFractions)
colnames(celFrac_ilr) = paste0("cellFrac_ilr_", 1:3)

METADATA = merge(METADATA, df_cellFractions, by.x = 'SampleID', by.y='row.names')
METADATA = merge(METADATA, celFrac_ilr, by.x = 'SampleID', by.y='row.names')
```


```{r preprocess.data1, results='asis'}
# Match covariates to expression data
indToRetain = intersect(METADATA$SampleID, colnames(geneCountsMerged))
     
geneCountsMerged = geneCountsMerged[,indToRetain]
     
rownames(METADATA) = METADATA$SampleID
```

## Summarize samples
Note that these numbers are based on Ethnicity, rather then genetically determined ancestry used in the filtering for eQTL analysis.
```{r summary}
METADATA %>% group_by(Dx, Institution, Region) %>% dplyr::summarise(count = n()) %>% Â©
  kable(caption="Diagnosis, Institution, Region") %>% kable_styling(full_width=FALSE) 
      
METADATA %>% group_by( Institution, Reported_Gender, Region) %>% dplyr::summarise(count = n()) %>% 
  kable(caption="Institution, Sex, Region") %>% kable_styling(full_width=FALSE) 

METADATA %>% group_by(Dx, Institution, Reported_Gender, Region) %>% dplyr::summarise(count = n()) %>% kable(caption="Diagnosis, Institution, Sex, Region") %>% kable_styling(full_width=FALSE) 

METADATA %>% group_by(Ethnicity, Institution, Region) %>% dplyr::summarise(count = n()) %>% kable(caption="Ethnicity, Institution, Region") %>% kable_styling(full_width=FALSE) 

METADATA %>% group_by(Ethnicity, Cohort, Region) %>% dplyr::summarise(count = n()) %>%  kable(caption="Ethnicity, Cohort, Region") %>% kable_styling(full_width=FALSE) 

METADATA %>% group_by(Cohort, Region) %>% dplyr::summarise(count = n()) %>%  kable(caption="Cohort, Region") %>% kable_styling(full_width=FALSE) 

METADATA %>% group_by(Cohort, Region, Ethnicity) %>% dplyr::summarise(count = n()) %>%  kable(caption="Cohort, Region") %>% kable_styling(full_width=FALSE) 
```



```{r}
CODE <- Folder(name = "Evaluate PEER in each cohort and brain region", parentId = "syn22012942")
CODE <- synStore(CODE)

# Store data 
storeData = function( data, file, description){
  saveRDS( data, file=file)

  nEXP_OBJ = File(file, 
                  name = description, 
                  parentId = CODE$properties$id)
  synStore(nEXP_OBJ, used = ALL_USED_IDs, executed = thisFile)
}
```
 

# Normalize and process
Each region and chort separately
```{r process.counts}
# keep genes that are express in at least one combination of cohort and gender
isexpr.lst = dlply(METADATA, .(Cohort, Region), .fun = function(mtd){
  rowSums(cpm(geneCountsMerged[,mtd$SampleID])>.5) >= .3*ncol(geneCountsMerged[,mtd$SampleID])
  })

isexpr = do.call(rbind, isexpr.lst)
isexpr = colSums(isexpr) > 0 # keep gene expressed in both cohorts 
table(isexpr)
   
grid = expand.grid(levels(METADATA$Cohort), unique(METADATA$Region))

# Normalize each cohort separately
dge.lst = lapply( 1:nrow(grid), function(i){

  Cohort = grid$Var1[i]
  Region = grid$Var2[i]

  idx = (METADATA$Cohort==Cohort)&(METADATA$Region == Region)

  cat(i, sum(idx), '\n')
  if( sum(idx) == 0){
    res = NULL
  }else{
    # normalize counts
    dge = DGEList( geneCountsMerged[isexpr,idx] )
    res = calcNormFactors(dge, "TMM")
  }
  res
})
names(dge.lst) = apply(grid, 1, paste, collapse=' - ')
dge.lst = dge.lst[!sapply(dge.lst, is.null)]


# formula used to fit model
form.lst = list(
  `MSSM-Penn-Pitt`  = ~ Dx + Reported_Gender + RIN +scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + Institution*(ageOfDeath + cellFrac_ilr_1 + cellFrac_ilr_2 + cellFrac_ilr_3 + EV.1 + EV.2 + EV.3 + EV.4 + EV.5),
  `NIMH-HBCC`       = ~ Dx + Reported_Gender + RIN + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + ageOfDeath + cellFrac_ilr_1 + cellFrac_ilr_2 + cellFrac_ilr_3 + EV.1 + EV.2 + EV.3 + EV.4 + EV.5)

for( datID in names(dge.lst) ){

  # get log2 CPM gene expression
  dge = dge.lst[[datID]]

  # get metadata in the same order
  i = match(colnames(dge), rownames(METADATA))
  info = METADATA[i,]

  # get formula
  j = unlist(sapply(names(form.lst), function(x) grep(x, datID)))
  form_use = form.lst[[names(j)]]
  # apply voom
  vobj = voomWithDreamWeights(dge, form_use, info, save.plot=TRUE, BPPARAM = SerialParam())

  # construct covariate matrix
  cov_matrix = model.matrix(form_use, info)

  # fit PEER
  model = PEER()

  PEER_setPhenoMean(model, t(vobj$E))

  PEER_setCovariates(model, cov_matrix)

  PEER_setNk(model,50)

  PEER_update(model)

  # extract PEER latent variables
  # first ncol(cov_matrix) columns are the oringal covariates
  latentVariables = PEER_getX(model)[,-c(1:ncol(cov_matrix))]
  rownames(latentVariables) = rownames(cov_matrix)
  colnames(latentVariables) = paste0("PEER_", 1:ncol(latentVariables))

  # upload observed expression
  file = paste0('geneExpr_',gsub(" ", '', gsub("-", "_", datID)), ".RDS")
  storeData( vobj$E, file, paste0(datID, " Observed expression as log2 CPM"))

  # upload covariate matrix
  file = paste0('covariateMatrix_',gsub(" ", '', gsub("-", "_", datID)), ".RDS")
  storeData( cov_matrix, file, paste0(datID, " Covariates for QTL analysis"))

  # upload latent variables
  file = paste0('PEER',gsub(" ", '', gsub("-", "_", datID)), ".RDS")
  storeData( latentVariables, file, paste0(datID, " Latent variables for QTL analysis"))
}
```

For each cohort and each brain region, compute TMM normalization, log2 CPM with voomWithDreamWeights, and a set of covariates.  Then compute PEER components using the observed expression and the covariates

`MSSM-Penn-Pitt`  = ~ Dx + Reported_Gender + RIN +scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + Institution*(ageOfDeath + cellFrac_ilr_1 + cellFrac_ilr_2 + cellFrac_ilr_3) + EV.1 + EV.2 + EV.3 + EV.4 + EV.5

`NIMH-HBCC`       = ~ Dx + Reported_Gender + RIN + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + ageOfDeath + cellFrac_ilr_1 + cellFrac_ilr_2 + cellFrac_ilr_3 + EV.1 + EV.2 + EV.3 + EV.4 + EV.5


# Cell Type Specific expression from bMIND

```{r create.folder.2}
# create new parent folder for these results
CODE <- Folder(name = "CTS - Evaluate PEER in each cohort and brain region", parentId = "syn23567522")
CODE <- synStore(CODE)

# Store data 
storeData = function( data, file, description){
  saveRDS( data, file=file)

  nEXP_OBJ = File(file, 
                  name = description, 
                  parentId = CODE$properties$id)
  # synStore(nEXP_OBJ, used = ALL_USED_IDs, executed = thisFile)
}
```
 

```{r CTS.download}
geneExpr.CTS = readRDS( synGet('syn23234833')$path )
ALL_USED_IDs = c(ALL_USED_IDs, 'syn23234833')       
```

```{r CTS.eval_peer}
# formula used to fit model
form.lst = list(
  `MSSM-Penn-Pitt`  = ~ Dx + Reported_Gender + RIN +scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + Institution*ageOfDeath,
  `NIMH-HBCC`       = ~ Dx + Reported_Gender + RIN + scale(IntronicRate) + scale(IntragenicRate) + scale(IntergenicRate) + scale(rRNARate) + ageOfDeath)

# for each dataset
for( datID in names(dge.lst) ){

  # get log2 CPM gene expression
  dge = dge.lst[[datID]]
  
  # get metadata in the same order
  i = match(colnames(dge), rownames(METADATA))
  info = METADATA[i,]

  # get formula
  j = unlist(sapply(names(form.lst), function(x) grep(x, datID)))
  form_use = form.lst[[names(j)]]

  # extract CTS data corresponding to info
  idx = match(info$SampleID, dimnames(geneExpr.CTS)[[3]])
  geneExpr = geneExpr.CTS[,,idx]

  # identical(dimnames(geneExpr)[[3]], info$SampleID)

  # for each cell type
  for( i in 1:length(dimnames(geneExpr)[[2]]) ){

    CT = dimnames(geneExpr)[[2]][i]

    # construct covariate matrix
    cov_matrix = model.matrix(form_use, info)

    # fit PEER
    model = PEER()

    PEER_setPhenoMean(model, t(geneExpr[,i,]))

    PEER_setCovariates(model, cov_matrix)

    PEER_setNk(model,2)

    PEER_update(model)

    # extract PEER latent variables
    # first ncol(cov_matrix) columns are the oringal covariates
    latentVariables = PEER_getX(model)[,-c(1:ncol(cov_matrix))]
    rownames(latentVariables) = rownames(cov_matrix)
    colnames(latentVariables) = paste0("PEER_", 1:ncol(latentVariables))

    # upload observed expression
    file = paste0('geneExpr_',gsub(" ", '', gsub("-", "_", datID)), '_', CT,".RDS")
    storeData( geneExpr, file, paste0(datID, " Observed expression", ' - ', CT))

    # upload covariate matrix
    file = paste0('covariateMatrix_',gsub(" ", '', gsub("-", "_", datID)), '_', CT, ".RDS")
    storeData( cov_matrix, file, paste0(datID, " Covariates for QTL analysis", ' - ' , CT))

    # upload latent variables
    file = paste0('PEER_',gsub(" ", '', gsub("-", "_", datID)), '_', CT, ".RDS")
    storeData( latentVariables, file, paste0(datID, " Latent variables for QTL analysis", ' - ', CT))
  }
}
```

for FILE in $(ls PEER*)
do
  mv $FILE $(echo $FILE | sed 's/PEER/PEER_/')
done


























<!---

  # add these terms back into the model
  form_add = ifelse(names(j) == "MSSM-Penn-Pitt",
        '~ Dx + Reported_Gender + Institution*(cellFrac_ilr_1 + cellFrac_ilr_2 + cellFrac_ilr_3)',
       ' ~ Dx + Reported_Gender + cellFrac_ilr_1 + cellFrac_ilr_2 + cellFrac_ilr_3')
  form_add = as.formula(form_add)



  # Compute residuals, but add back in Dx*Reported_Gender
  # for form = ~ Dx*Reported_Gender
  f = function(fit){
    residuals(fit) + get_prediction( fit, form_add)
  }

  # Compute residuals while adding back biological variables
  resid.lst = fitVarPartModel(vobj, form_use, info, fxn = f, BPPARAM = SnowParam(18))  
  residMat = do.call(rbind, resid.lst)

   # upload residuals
  file = paste0('metadata_',gsub(" ", '', gsub("-", "_", datID)), ".RDS")
  storeData( residMat, file, paste0(datID, " Residuals removing technical variables"))

  # upload covariates
  info.sub = info[,all.vars(form_use)]
  file = paste0('metadata_',gsub(" ", '', gsub("-", "_", datID)), ".RDS")
  storeData( info.sub, file, paste0(datID, " Meta data for variables relavant for QTL analysis"))

  })
names(residMat.lst) = names(dge.lst)
--->



















