---
title: "Sex-specific Meta-analysis in MSSM-Penn-Pitt and HBCC ACC/DLPFC"
author: "KM and GH"
date: "`r date()`"
output: html_document
---

```{r knit2synapse, eval=FALSE}
# rmarkdown::render("graphics_t_statistics.Rmd")
library(synapser)
library(knit2synapse)
synLogin()
knit2synapse::createAndKnitToFolderEntity(file = "individual-cohorts/graphics_t_statistics.Rmd",
                                          parentId ="syn21435176",
                                          folderName = 'Sex-specific Meta-analysis')
```

```{r setup, include=FALSE}
library(synapser)
synLogin()
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(dplyr)
library(ComplexHeatmap)
library(readxl)
library(gtools)
library(corrplot)
library(pvclust)

synLogin()

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE, eval=FALSE}
thisFileName <- 'graphics_t_statistics.Rmd'

# Github link
thisRepo <- getRepo(repository = "CommonMindConsortium/covarr-de", ref="branch", refName='master')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('individual-cohorts/',thisFileName))
```

```{r multiplot}
# Multiple plot function
#
# ggplot objects can be passed in ..., or to plotlist (as a list of ggplot objects)
# - cols:   Number of columns in layout
# - layout: A matrix specifying the layout. If present, 'cols' is ignored.
#
# If the layout is something like matrix(c(1,2,3,3), nrow=2, byrow=TRUE),
# then plot 1 will go in the upper left, 2 will go in the upper right, and
# 3 will go all the way across the bottom.
#
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)

  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)

  numPlots = length(plots)

  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                    ncol = cols, nrow = ceiling(numPlots/cols))
  }

 if (numPlots==1) {
    print(plots[[1]])

  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))

    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))

      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}
```


## t-statistics from differential expression results

- separate by comparison

```{r data}
mpp_acc <- readr::read_tsv(synGet("syn22079117")$path) %>% 
  dplyr::select(Comparison, gene_id, AveExpr, t, logFC, P.Value, adj.P.Val) %>% 
  dplyr::rename(`MSSM-Penn-Pitt ACC t-statistic` = t)
mpp_dlpfc <- readr::read_tsv(synGet("syn22079472")$path) %>% 
  dplyr::select(Comparison, gene_id, AveExpr, t, logFC, P.Value, adj.P.Val) %>% 
  dplyr::rename(`MSSM-Penn-Pitt DLPFC t-statistic` = t)
hbcc_dsACC <- readr::read_tsv(synGet("syn22086829")$path) %>% 
  dplyr::select(Comparison, gene_id, AveExpr, t, logFC, P.Value, adj.P.Val) %>% 
  dplyr::rename(`HBCC dsACC t-statistic` = t)
hbcc_stg <- readr::read_tsv(synGet("syn22087386")$path) %>% 
  dplyr::select(Comparison, gene_id, AveExpr, t) %>% 
  dplyr::rename(`HBCC STG t-statistic` = t)
hbcc_sgACC <- readr::read_tsv(synGet("syn22085862")$path) %>% 
  dplyr::select(Comparison, gene_id, AveExpr, t, logFC, P.Value, adj.P.Val) %>% 
  dplyr::rename(`HBCC sgACC t-statistic` = t)
hbcc_dlpfc <- readr::read_tsv(synGet("syn22084205")$path) %>% 
  dplyr::select(Comparison, gene_id, AveExpr, t, logFC, P.Value, adj.P.Val) %>% 
  dplyr::rename(`HBCC DLPFC t-statistic` = t)
# CMC Analysis 1
CMC1 <- readr::read_tsv(synGet("syn5607603")$path) %>% 
  dplyr::rename(gene_id = genes,
                `CMC 1 t-statistic` = t) %>% 
  dplyr::select(gene_id, `CMC 1 t-statistic`)
# HBCC-DLPFC microarray
CMC_HBCC <- readr::read_tsv(synGet("syn4941553")$path) %>% 
  dplyr::rename(gene_id = genes,
                `HBCC microarray t-statistic` = t) %>% 
  dplyr::select(gene_id, `HBCC microarray t-statistic`)
# Gandal, Science 2018, Table S1
# Use Tab:DGE
# wget https://science.sciencemag.org/highwire/filestream/720110/field_highwire_adjunct_files/1/aat8127_Table_S1.xlsx
gandal <- readxl::read_xlsx( synGet("syn22255217")$path, sheet = "DGE") %>% 
  dplyr::rename(gene_id = ensembl_gene_id,
                `Gandal, SCZ t-statistic` = `SCZ.t.value`,
                `Gandal, BD t-statistic` = `BD.t.value`,
                `Gandal, ASD t-statistic` = `ASD.t.value`) %>%
  dplyr::select(gene_id, `Gandal, SCZ t-statistic`,
                `Gandal, BD t-statistic`, `Gandal, ASD t-statistic`)

# AMP-AD RNASeq reprocessing GRCh37 - https://www.synapse.org/#!Synapse:syn9702085
MSSM <- readr::read_tsv(synGet("syn10526259")$path) %>%
  dplyr::filter(Comparison %in% "AD-CONTROL",
                Model %in% "Diagnosis") %>%
  dplyr::select(ensembl_gene_id, Tissue, t) %>%
  dplyr::group_by(ensembl_gene_id, Tissue) %>% 
  dplyr::filter(row_number() == 1) %>% 
  tidyr::pivot_wider(names_from = Tissue,
                     values_from = t) %>%
  dplyr::rename(`AMP-AD MSSM FP AD t-statistic` = FP,
                `AMP-AD MSSM IGF AD t-statistic` = IFG,
                `AMP-AD MSSM PHG AD t-statistic` = PHG,
                `AMP-AD MSSM STG AD t-statistic` = STG,
                gene_id = ensembl_gene_id) %>% 
  dplyr::select(gene_id, `AMP-AD MSSM STG AD t-statistic`) 
MSSM <- MSSM[!is.na(MSSM$`AMP-AD MSSM STG AD t-statistic`),]
  
MAYO <- readr::read_tsv(synGet("syn8468023")$path) %>% 
  dplyr::filter(Comparison %in% "AD-CONTROL",
                Model %in% "Diagnosis") %>%
  dplyr::select(ensembl_gene_id, `Tissue.ref`, t) %>%
  dplyr::group_by(ensembl_gene_id, `Tissue.ref`) %>% 
  dplyr::filter(row_number() == 1) %>% 
  tidyr::pivot_wider(names_from = `Tissue.ref`,
                     values_from = t) %>% 
  dplyr::rename(`AMP-AD MAYO CBE AD t-statistic` = CBE,
                `AMP-AD MAYO TCX AD t-statistic` = TCX,
                gene_id = ensembl_gene_id) %>% 
  dplyr::select(-`AMP-AD MAYO TCX AD t-statistic`)
  
ROSMAP <- readr::read_tsv(synGet("syn8456721")$path) %>%
    dplyr::filter(Comparison %in% "AD-CONTROL",
                Model %in% "Diagnosis") %>%
  dplyr::rename(gene_id = ensembl_gene_id, 
                `AMP-AD ROSMAP DLPFC AD t-statistic` = t) %>% 
  dplyr::select(gene_id, `AMP-AD ROSMAP DLPFC AD t-statistic`)
  
out_group <- inner_join(MSSM, MAYO) %>% 
  inner_join(., ROSMAP) %>% 
  inner_join(., gandal)
  
df <- inner_join(mpp_acc, mpp_dlpfc,
                 by = c("Comparison", "gene_id"))
df <- inner_join(df, hbcc_dsACC, by = c("Comparison", "gene_id"))
df <- inner_join(df, hbcc_stg, by = c("Comparison", "gene_id"))
df <- inner_join(df, hbcc_sgACC, by = c("Comparison", "gene_id"))
df <- inner_join(df, hbcc_dlpfc, by = c("Comparison", "gene_id"))
```

```{r functions}
plot_t <- function(df, conditions){
  p = list()
  for(i in seq_along(conditions[,1])){
  positions = conditions[i,]
  x = df[positions[1]]
  y = df[positions[2]]
  computed_cor = cor(x, y, method = "spearman")
  x_var <- names(x)
  y_var <- names(y)
  p[[i]] = ggplot(df, aes(x = .data[[x_var]], y = .data[[y_var]])) +
  geom_hex() +
  geom_abline() +
  geom_smooth(method = "lm") +
  theme_bw() +
  xlim(8,-8) + 
  ylim(8,-8) +
    labs(
    caption = paste0("spearman correlation between the statistics is ", round(computed_cor, digits = 3))
  ) +
    theme(aspect.ratio=1, 
          axis.text=element_text(size=12),
          plot.caption = element_text(face = "bold", size = "12"))
  }
multiplot(plotlist = p, cols = 2)
}
```

# Control vs Other
```{r plota, fig.width = 10, fig.height = 100}
df_one <- df[df$Comparison %in% "Control_vs_Other",] %>% 
  dplyr::select_if( ~ !all(is.na(.))) %>% 
  dplyr::select(-Comparison)

#Remove transcript Ids
df_one <- df_one %>% 
  tidyr::separate(gene_id, c("gene_id", "transcript")) %>% 
  dplyr::select(-one_of("transcript"))
df_one <- inner_join(df_one, out_group, by = c("gene_id")) %>% 
  group_by(gene_id) %>% 
  dplyr::filter(row_number()==1) %>% 
  tibble::column_to_rownames(var = "gene_id")

conditions <- gtools::combinations(n = ncol(df_one), 2, seq_len(ncol(df_one)))
plot_t(df_one, conditions)
```

```{r heatmapa, fig.width = 8, fig.height = 8 }
colnames(df_one) <- gsub(" t-statistic", "", colnames(df_one))

cor <- cor(df_one)
# clustering based on distance
hcl = hclust(as.dist(1-cor))
# reorder based on clustering
reorder = cor[hcl$order, hcl$order]
results <- pvclust(cor)
# plot clustering
plot(results, hang=-1)
col2 <- colorRampPalette(rev(c("#67001F", "#B2182B", "#D6604D", "#F4A582",
                           "#FDDBC7", "#FFFFFF", "#D1E5F0", "#92C5DE",
                           "#4393C3", "#2166AC", "#053061")))
corrplot(reorder, col = col2(200))
```


# Control vs SCZ
```{r plotb, fig.width = 10, fig.height = 150}
df_two <- df[df$Comparison %in% "Control_vs_SCZ",] %>% 
  dplyr::select_if( ~ !all(is.na(.))) %>% 
  dplyr::select(-Comparison)

#Remove transcript Ids
df_two <- df_two %>% 
  tidyr::separate(gene_id, c("gene_id", "transcript")) %>% 
  dplyr::select(-one_of("transcript"))
df_two <- inner_join(df_two, CMC1, by = c("gene_id")) 
df_two <- inner_join(df_two, CMC_HBCC, by = c("gene_id"))
df_two <- inner_join(df_two, out_group, by = c("gene_id")) %>% 
  group_by(gene_id) %>% 
  dplyr::filter(row_number()==1) %>% 
  tibble::column_to_rownames(var = "gene_id")

conditions <- gtools::combinations(n = ncol(df_two), 2, seq_len(ncol(df_two)))
plot_t(df_two, conditions)
```


```{r heatmapb, fig.width = 8, fig.height = 8 }
colnames(df_two) <- gsub(" t-statistic", "", colnames(df_two))
  
cor <- cor(df_two)
# clustering based on distance
hcl = hclust(as.dist(1-cor))
# reorder based on clustering
reorder = cor[hcl$order, hcl$order]
results <- pvclust(cor)
# plot clustering
plot(results, hang=-1)
plot(hcl, hang=-1)
corrplot(reorder, col = col2(200))
```

```{r custom.functions}

library(metafor)

# perform meta-analysis given fits from limma
run_meta_analysis = function( tabList, coef, method="REML" ){

  if( length(coef) > 1){
    stop("Only supports one coefficient term")
  }

  # extract logFC and se for each model fit
  resList = lapply( tabList, function(tab){
    tab$se = tab$logFC /  tab$t
    tab
  })

  # extract features names
  featureNames = lapply(resList, function(res){
    rownames(res)
    })
  featureNames = unique(unlist(featureNames))

  # check that each fit has all features
  # testIdentical = lapply(resList, function(res){
  #   identical(sort(rownames(res)), sort(featureNames))
  #   })
  # if(  any(!unlist(testIdentical)) ){
  #   stop("each entry in fitList must have all features with the same names")
  # }

  # apply meta analysis for each feature
  resRMA = lapply(featureNames, function(feature){
    # extract summary statistics for this feature from all studies
    res = lapply(resList, function(res){
      res[feature,]
      })
    df = do.call("rbind", res)
    suppressWarnings(rma(yi=logFC, sei=se, data = df, method=method))
    })
  names(resRMA) = featureNames

  # extract results
  # Currently exatract simple results
  # other results may be relevant if meta-analysing more than two studies
  resTable = lapply(resRMA, function(x){
    data.frame( logFC     = x$beta,
                se        = x$se,
                P.Value   = x$pval,
                adj.P.Val = p.adjust(x$pval, "fdr"),
                Q         = x$QE,
                N         = x$k,
                I2        = x$I2)
    })
  do.call("rbind", resTable)
}

plotVolcano = function(df, nshow=10){

  library(org.Hs.eg.db)
  library(ggrepel)

  # indicate significant genes
  df$isSignif = c("no","yes")[(df$adj.P.Val < 0.05)+1]

  # show gene names
  df$EnsID = gsub('\\..*$', '', rownames(df))

  df_hugo = select(org.Hs.eg.db, df$EnsID, "SYMBOL", "ENSEMBL")
  df = merge(df, df_hugo, by.x='EnsID', by.y="ENSEMBL")

  # df = df[!duplicated(df$gene_id),]

  # top significant genes
  df2 = df[df$adj.P.Val <= sort(df$adj.P.Val )[nshow], ]

  ggplot(df, aes( logFC, -log10(P.Value), color=isSignif)) + geom_point() + theme_bw(15) + theme(aspect.ratio=1, legend.position="none", plot.title = element_text(hjust = 0.5)) + xlab(bquote(log[2]~fold~change)) + ylab(bquote(-log[10]~P)) + scale_color_manual(values=c("grey", "red")) + geom_text_repel(data=df2, aes(logFC, -log10(P.Value), label=SYMBOL), segment.size=.5,  segment.color="black", color="black", force=1, nudge_x=.005, nudge_y=.5)
}


plot_forrest = function( resList, ensID){ 

  library(patchwork)

  # create a single data.frame with logFC and se
  df_forrest = lapply(resList, function(x)x[ensID,])
  df_forrest = do.call(rbind, df_forrest)
  df_forrest$Dataset = rownames(df_forrest)
  df_forrest$se = df_forrest$logFC /  df_forrest$t

  # bind only shared columns
  df = gtools::smartbind(df_forrest,  res_dlpfc[ensID,])
  df$Dataset = factor(df$Dataset, rev(df$Dataset))

  # set range to include zero
  rng = range(c(with(df, logFC-se), with(df, logFC+se)))
  if( rng[1] > 0 ){
    rng[1] = 0
  }
  if( rng[2] < 0 ){
    rng[2] = 0
  }

  # forrest plot
  fig1 = ggplot(df, aes(Dataset, logFC)) + geom_point(color="navy") + geom_errorbar(aes(ymin = logFC - se, ymax=logFC + se), width=0.1) + theme_bw(12) + theme(aspect.ratio= .2, plot.title = element_text(hjust = 0.5)) + coord_flip() + ylim(rng) + xlab(bquote(log[2]~fold~change)) + ggtitle('RCSD1')

  # Adjusted p-value
  ymax = max(-log10(df$adj.P.Val))*1.05
  fig2 = ggplot(df, aes(Dataset, -log10(adj.P.Val))) + geom_bar(stat="identity", fill="navy") + theme_bw(12) + theme( aspect.ratio= .5, plot.margin = unit(c(0,0,0,0), "cm"),
    axis.title.y=element_blank(),
          axis.text.y=element_blank(),
          axis.ticks.y=element_blank()) + coord_flip() + ylab(bquote(-log[10]~adjusted~P)) + scale_y_continuous(expand=c(0,0), limits=c(0, ymax))

  # plot_grid(fig1, fig2, align='vh')
  fig1 + fig2
}
```


# Violin plots
```{r make.volcano}

df_mpp_dlpfc <- readr::read_tsv(synGet("syn22079472")$path) %>% 
  dplyr::select(Comparison, gene_id, AveExpr, t, logFC, P.Value, adj.P.Val) %>% 
  dplyr::filter(Comparison == 'Control_vs_SCZ') %>% 
  dplyr::rename(`MSSM-Penn-Pitt DLPFC t-statistic` = t)  %>% 
  as.data.frame  %>% unique
rownames(df_mpp_dlpfc) = df_mpp_dlpfc$gene_id

plotVolcano( df_mpp_dlpfc, 15) + ggtitle("MSSM-Penn-Pitt - DLPFC")
```


```{r meta_analysis.SZ}

dlpfcList = list(MPP  = mpp_dlpfc %>% unique %>%
                        dplyr::filter( Comparison == 'Control_vs_SCZ') %>%
                        dplyr::rename(t = `MSSM-Penn-Pitt DLPFC t-statistic`) %>%
                        as.data.frame,
                 HBCC = hbcc_dlpfc %>%  unique %>%
                        dplyr::filter( Comparison == 'Control_vs_SCZ') %>% 
                        dplyr::rename(t = `HBCC DLPFC t-statistic`) %>%
                        as.data.frame)

rownames(dlpfcList[['MPP']]) = dlpfcList[['MPP']]$gene_id
rownames(dlpfcList[['HBCC']]) = dlpfcList[['HBCC']]$gene_id

res_dlpfc = run_meta_analysis( dlpfcList, "FE")
res_dlpfc$Dataset = "Meta-analysis"


accList = list(MPP  = mpp_acc %>% unique %>%
                        dplyr::filter( Comparison == 'Control_vs_SCZ') %>%
                        dplyr::rename(t = `MSSM-Penn-Pitt ACC t-statistic` )%>%
                        as.data.frame,
              HBCC  = hbcc_dsACC %>% unique %>%
                        dplyr::filter( Comparison == 'Control_vs_SCZ') %>% 
                        dplyr::rename(t = `HBCC dsACC t-statistic`) %>%
                        as.data.frame)

rownames(accList[['MPP']]) = accList[['MPP']]$gene_id
rownames(accList[['HBCC']]) = accList[['HBCC']]$gene_id

res_acc = run_meta_analysis( accList, "FE")
res_acc$Dataset = "Meta-analysis"
```

```{r plot.meta_analysis}

library(cowplot)

table(res_dlpfc$adj.P.Val < 0.05)
table(res_acc$adj.P.Val < 0.05)


# forrest plot of meta-analysis
# This is a bit of a mess
ensID = 'ENSG00000198771.11'
res_dlpfc[ensID,]

lapply(dlpfcList, function(x) x[ensID,])

plotVolcano( res_dlpfc, 20) + ggtitle("Meta analysis - DLPFC")


ensID = rownames(res_dlpfc)[which.min(res_dlpfc$adj.P.Val)][1]

plot_forrest(dlpfcList, ensID)

plotVolcano( res_acc, 20) + ggtitle("Meta analysis - ACC")
```


# Other vs SCZ
```{r plotc, fig.width = 10, fig.height = 100}
df_three <- df[df$Comparison %in% "Other_vs_SCZ",] %>% 
  dplyr::select_if( ~ !all(is.na(.))) %>% 
  dplyr::select(-Comparison)

#Remove transcript Ids
df_three <- df_three %>% 
  tidyr::separate(gene_id, c("gene_id", "transcript")) %>% 
  dplyr::select(-one_of("transcript"))
df_three <- inner_join(df_three, out_group, by = c("gene_id")) %>% 
  group_by(gene_id) %>% 
  dplyr::filter(row_number()==1) %>% 
  tibble::column_to_rownames(var = "gene_id")

conditions <- gtools::combinations(n = ncol(df_three), 2, seq_len(ncol(df_three)))
plot_t(df_three, conditions)
```


```{r heatmapc, fig.width = 8, fig.height = 8 }
colnames(df_three) <- gsub(" t-statistic", "", colnames(df_three))
  
cor <- cor(df_three)
# clustering based on distance
hcl = hclust(as.dist(1-cor))
# reorder based on clustering
reorder = cor[hcl$order, hcl$order]
results <- pvclust(cor)
# plot clustering
plot(results, hang=-1)
corrplot(reorder, col = col2(200))
```

Code: r thisFile

```{r store.files, eval=FALSE}
CODE <- Folder(name = "Sex-specific Meta-analysis", parentId = "syn21435176")
CODE <- synStore(CODE)
```
