---
title: "Graphics"
author: "KM"
date: "`r date()`"
output: html_document
---

```{r knit2synapse, eval=FALSE}
library(synapser)
library(knit2synapse)
synLogin()
knit2synapse::createAndKnitToFolderEntity(file = "individual-cohorts/graphics_t_statistics.Rmd",
                                          parentId ="syn22012942",
                                          folderName = 't-statistics across cohorts')
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(dplyr)
library(ComplexHeatmap)
library(readxl)
library(gtools)

synLogin()

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
thisFileName <- 'graphics_t_statistics.Rmd'

# Github link
thisRepo <- getRepo(repository = "CommonMindConsortium/covarr-de", ref="branch", refName='master')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('individual-cohorts/',thisFileName))
```

## t-statistics from differential expression results

- separate by comparison

```{r data}
mpp_acc <- readr::read_tsv(synGet("syn22079117")$path) %>% 
  dplyr::select(Comparison, gene_id, t) %>% 
  dplyr::rename(`MSSM-Penn-Pitt ACC t-statistic` = t)
mpp_dlpfc <- readr::read_tsv(synGet("syn22079472")$path) %>% 
  dplyr::select(Comparison, gene_id, t) %>% 
  dplyr::rename(`MSSM-Penn-Pitt DLPFC t-statistic` = t)
hbcc_dsACC <- readr::read_tsv(synGet("syn22086829")$path) %>% 
  dplyr::select(Comparison, gene_id, t) %>% 
  dplyr::rename(`HBCC dsACC t-statistic` = t)
hbcc_stg <- readr::read_tsv(synGet("syn22087386")$path) %>% 
  dplyr::select(Comparison, gene_id, t) %>% 
  dplyr::rename(`HBCC STG t-statistic` = t)
hbcc_sgACC <- readr::read_tsv(synGet("syn22085862")$path) %>% 
  dplyr::select(Comparison, gene_id, t) %>% 
  dplyr::rename(`HBCC sgACC t-statistic` = t)
hbcc_dlpfc <- readr::read_tsv(synGet("syn22084205")$path) %>% 
  dplyr::select(Comparison, gene_id, t) %>% 
  dplyr::rename(`HBCC DLPFC t-statistic` = t)
# CMC Analysis 1
CMC1 <- readr::read_tsv(synGet("syn5607603")$path) %>% 
  dplyr::rename(gene_id = genes,
                `CMC 1 t-statistic` = t) %>% 
  dplyr::select(gene_id, `CMC 1 t-statistic`)
# HBCC-DLPFC microarray
CMC_HBCC <- readr::read_tsv(synGet("syn4941553")$path) %>% 
  dplyr::rename(gene_id = genes,
                `HBCC microarray t-statistic` = t) %>% 
  dplyr::select(gene_id, `HBCC microarray t-statistic`)
# Gandal, Science 2018, Table S1
# Use Tab:DGE
# wget https://science.sciencemag.org/highwire/filestream/720110/field_highwire_adjunct_files/1/aat8127_Table_S1.xlsx
gandal <- readxl::read_xlsx('../aat8127_Table_S1.xlsx', sheet = "DGE") %>% 
  dplyr::rename(gene_id = ensembl_gene_id,
                `Gandal, SCZ t-statistic` = `SCZ.t.value`,
                `Gandal, BD t-statistic` = `BD.t.value`,
                `Gandal, ASD t-statistic` = `ASD.t.value`) %>%
  dplyr::select(gene_id, `Gandal, SCZ t-statistic`,
                `Gandal, BD t-statistic`, `Gandal, ASD t-statistic`)
  

df <- inner_join(mpp_acc, mpp_dlpfc,
                 by = c("Comparison", "gene_id"))
df <- inner_join(df, hbcc_dsACC, by = c("Comparison", "gene_id"))
df <- inner_join(df, hbcc_stg, by = c("Comparison", "gene_id"))
df <- inner_join(df, hbcc_sgACC, by = c("Comparison", "gene_id"))
df <- inner_join(df, hbcc_dlpfc, by = c("Comparison", "gene_id"))
```

```{r functions}
plot_t <- function(df, conditions){
  p = list()
  for(i in seq_along(conditions[,1])){
  positions = conditions[i,]
  x = df[positions[1]]
  y = df[positions[2]]
  computed_cor = cor(x, y, method = "spearman")
  x_var <- names(x)
  y_var <- names(y)
  p[[i]] = ggplot(df, aes(x = .data[[x_var]], y = .data[[y_var]])) +
  geom_hex() +
  geom_abline() +
  theme_bw() +
  labs(
    caption = paste0("spearman correlation between the statistics is ", round(computed_cor, digits = 3))
  ) +
    theme(aspect.ratio=1, 
          axis.text=element_text(size=12),
          plot.caption = element_text(face = "bold", size = "12"))
  }
multiplot(plotlist = p, cols = 2)
}
```

# Control vs Other
```{r plota, fig.width = 10, fig.height = 100}
df_one <- df[df$Comparison %in% "Control_vs_Other",] %>% 
  dplyr::select_if( ~ !all(is.na(.))) %>% 
  dplyr::select(-Comparison)

#Remove transcript Ids
df_one <- df_one %>% 
  separate(gene_id, c("gene_id", "transcript")) %>% 
  dplyr::select(-one_of("transcript"))
df_one <- inner_join(df_one, dplyr::select(gandal, gene_id,  `Gandal, SCZ t-statistic`,`Gandal, BD t-statistic`,`Gandal, ASD t-statistic`), by = c("gene_id")) %>% 
  group_by(gene_id) %>% 
  dplyr::filter(row_number()==1) %>% 
  tibble::column_to_rownames(var = "gene_id")

conditions <- gtools::combinations(n = ncol(df_one), 2, seq_len(ncol(df_one)))
plot_t(df_one, conditions)
```

```{r heatmapa, fig.width = 8, fig.height = 8 }
colnames(df_one) <- gsub(" t-statistic", "", colnames(df_one))
  
cor <- corr.test(df_one, use = "pairwise.complete.obs", method = "spearman")

col2 <- colorRampPalette(rev(c("#67001F", "#B2182B", "#D6604D", "#F4A582",
                           "#FDDBC7", "#FFFFFF", "#D1E5F0", "#92C5DE",
                           "#4393C3", "#2166AC", "#053061")))
corrplot(cor$r, col = col2(200))
```


# Control vs SCZ
```{r plotb, fig.width = 10, fig.height = 100}
df_two <- df[df$Comparison %in% "Control_vs_SCZ",] %>% 
  dplyr::select_if( ~ !all(is.na(.))) %>% 
  dplyr::select(-Comparison)

#Remove transcript Ids
df_two <- df_two %>% 
  separate(gene_id, c("gene_id", "transcript")) %>% 
  dplyr::select(-one_of("transcript"))
df_two <- inner_join(df_two, CMC1, by = c("gene_id")) 
df_two <- inner_join(df_two, CMC_HBCC, by = c("gene_id"))
df_two <- inner_join(df_two, dplyr::select(gandal, gene_id, `Gandal, SCZ t-statistic`, 
                                           `Gandal, ASD t-statistic`, `Gandal, BD t-statistic`), by = c("gene_id")) %>% 
  group_by(gene_id) %>% 
  dplyr::filter(row_number()==1) %>% 
  tibble::column_to_rownames(var = "gene_id")

conditions <- gtools::combinations(n = ncol(df_two), 2, seq_len(ncol(df_two)))
plot_t(df_two, conditions)
```


```{r heatmapb, fig.width = 8, fig.height = 8 }
colnames(df_two) <- gsub(" t-statistic", "", colnames(df_two))
  
cor <- corr.test(df_two, use = "pairwise.complete.obs", method = "spearman")

corrplot(cor$r, col = col2(200))
```

# Other vs SCZ
```{r plotc, fig.width = 10, fig.height = 100}
df_three <- df[df$Comparison %in% "Other_vs_SCZ",] %>% 
  dplyr::select_if( ~ !all(is.na(.))) %>% 
  dplyr::select(-Comparison)

#Remove transcript Ids
df_three <- df_three %>% 
  separate(gene_id, c("gene_id", "transcript")) %>% 
  dplyr::select(-one_of("transcript"))
df_three <- inner_join(df_three, dplyr::select(gandal, gene_id,  `Gandal, BD t-statistic`,
                                               `Gandal, ASD t-statistic`, `Gandal, SCZ t-statistic`), by = c("gene_id")) %>% 
  group_by(gene_id) %>% 
  dplyr::filter(row_number()==1) %>% 
  tibble::column_to_rownames(var = "gene_id")

conditions <- gtools::combinations(n = ncol(df_three), 2, seq_len(ncol(df_three)))
plot_t(df_three, conditions)
```


```{r heatmapc, fig.width = 8, fig.height = 8 }
colnames(df_three) <- gsub(" t-statistic", "", colnames(df_three))
  
cor <- corr.test(df_three, use = "pairwise.complete.obs", method = "spearman")

corrplot(cor$r, col = col2(200))
```

Code: `r thisFile`

```{r store.files}
CODE <- Folder(name = "t-statistics across cohorts", parentId = "syn22012942")
CODE <- synStore(CODE)
```
